{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(global) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  if (typeof (window._bokeh_onload_callbacks) === \"undefined\") {\n",
       "    window._bokeh_onload_callbacks = [];\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    window._bokeh_onload_callbacks.forEach(function(callback) { callback() });\n",
       "    delete window._bokeh_onload_callbacks\n",
       "    console.info(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(js_urls, callback) {\n",
       "    window._bokeh_onload_callbacks.push(callback);\n",
       "    if (window._bokeh_is_loading > 0) {\n",
       "      console.log(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.log(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    window._bokeh_is_loading = js_urls.length;\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      var s = document.createElement('script');\n",
       "      s.src = url;\n",
       "      s.async = false;\n",
       "      s.onreadystatechange = s.onload = function() {\n",
       "        window._bokeh_is_loading--;\n",
       "        if (window._bokeh_is_loading === 0) {\n",
       "          console.log(\"Bokeh: all BokehJS libraries loaded\");\n",
       "          run_callbacks()\n",
       "        }\n",
       "      };\n",
       "      s.onerror = function() {\n",
       "        console.warn(\"failed to load library \" + url);\n",
       "      };\n",
       "      console.log(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  var js_urls = ['https://cdn.pydata.org/bokeh/release/bokeh-0.11.1.min.js', 'https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.11.1.min.js', 'https://cdn.pydata.org/bokeh/release/bokeh-compiler-0.11.1.min.js'];\n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    \n",
       "    function(Bokeh) {\n",
       "      Bokeh.$(\"#41a6b671-7624-488f-93d6-efb752305186\").text(\"BokehJS successfully loaded\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-0.11.1.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-0.11.1.min.css\");\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.11.1.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-widgets-0.11.1.min.css\");\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    for (var i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i](window.Bokeh);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  if (window._bokeh_is_loading === 0) {\n",
       "    console.log(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(js_urls, function() {\n",
       "      console.log(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(this));"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import cameo\n",
    "from cobra import Model, Reaction, Metabolite\n",
    "from cobra.io import read_sbml_model\n",
    "from cobra.io import save_json_model\n",
    "from cameo.flux_analysis.simulation import pfba\n",
    "import cobra.test\n",
    "import os\n",
    "from Functions_Modules.curation_tools import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "relative_directory = os.getcwd()\n",
    "filename = relative_directory + '/Reconstructions/MethylococcusModel8.xml'\n",
    "model = cameo.load_model(filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Final 'clean-up' step. I'll get rid of all MNXR or MNXM tagged reactions or metabolites."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('MNXM21342_c', frozenset([<Reaction MNXR85991_c at 0x10dacb750>, <Reaction MNXM21342_out at 0x10dc36f50>]))\n",
      "('MNXM5285_c', frozenset([<Reaction MNXM5285_out at 0x10dc36fd0>]))\n",
      "('MNXM7549_c', frozenset([<Reaction MNXM7549_out at 0x10dc3f090>]))\n",
      "('MNXM4463_c', frozenset([<Reaction MNXM4463_out at 0x10dc3f110>]))\n",
      "('MNXM6386_c', frozenset([<Reaction MNXM6386_out at 0x10dc3f190>]))\n",
      "('MNXM2142_c', frozenset([<Reaction MNXM2142_out at 0x10dc3f590>]))\n",
      "('MNXM7231_c', frozenset([<Reaction MNXM7231_out at 0x10dc3f810>, <Reaction MNXR68614_c at 0x10dad7e50>]))\n",
      "('MNXM95609_c', frozenset([<Reaction MNXM95609_out at 0x10dc492d0>]))\n",
      "('MNXM96150_c', frozenset([<Reaction MNXM96150_out at 0x10dc499d0>, <Reaction MNXR85732_c at 0x10dae9950>]))\n",
      "('MNXM92688_c', frozenset([<Reaction MNXM92688_out at 0x10dc53290>, <Reaction MNXR59223_c at 0x10db0e0d0>]))\n",
      "('MNXM1740_c', frozenset([<Reaction MNXM1740_out at 0x10dc53390>, <Reaction MNXR73718_c at 0x10db0e310>]))\n",
      "('MNXM2120_c', frozenset([<Reaction MNXM2120_out at 0x10dc53410>, <Reaction MNXR16109_c at 0x10db0e6d0>]))\n",
      "('MNXM3563_c', frozenset([<Reaction MNXM3563_out at 0x10dc53490>]))\n",
      "('MNXM1682_c', frozenset([<Reaction MNXM1682_out at 0x10dc53690>]))\n",
      "('MNXM5588_c', frozenset([<Reaction MNXR85034_c at 0x10db33e10>, <Reaction MNXM5588_out at 0x10dc53910>]))\n",
      "('MNXM12690_c', frozenset([<Reaction MNXM12690_out at 0x10dc53b10>, <Reaction MNXR59619_c at 0x10db392d0>]))\n",
      "('MNXM1293_c', frozenset([<Reaction MNXM1293_out at 0x10dc53b90>]))\n",
      "('MNXM5412_c', frozenset([<Reaction MNXM5412_out at 0x10dc53c10>, <Reaction MNXR74776_c at 0x10db39450>]))\n",
      "('MNXM5511_c', frozenset([<Reaction MNXM5511_out at 0x10dc53c90>]))\n",
      "('MNXM3034_c', frozenset([<Reaction MNXM3034_out at 0x10dc53d90>, <Reaction MNXR6705_c at 0x10db39bd0>]))\n",
      "('MNXM6152_c', frozenset([<Reaction MNXM6152_out at 0x10dc53e10>]))\n",
      "('MNXM6177_c', frozenset([<Reaction MNXM6177_out at 0x10dc53e90>]))\n",
      "('MNXM1651_c', frozenset([<Reaction MNXR7517_c at 0x10db40110>, <Reaction MNXM1651_out at 0x10dc53f10>]))\n",
      "('MNXM1731_c', frozenset([<Reaction MNXM1731_out at 0x10dc53f90>]))\n",
      "('MNXM4764_c', frozenset([<Reaction MNXR15981_c at 0x10db40590>, <Reaction MNXM4764_out at 0x10dc5d050>]))\n",
      "('MNXM2354_c', frozenset([<Reaction MNXM2354_out at 0x10dc5d0d0>]))\n",
      "('MNXM2709_c', frozenset([<Reaction MNXR17615_c at 0x10db40710>, <Reaction MNXM2709_out at 0x10dc5d150>]))\n",
      "('MNXM114652_c', frozenset([<Reaction MNXM114652_out at 0x10dc5d1d0>]))\n",
      "('MNXM1289_c', frozenset([<Reaction MNXR5404_c at 0x10db46290>, <Reaction MNXM1289_out at 0x10dc5d250>, <Reaction MNXR35657_c at 0x10db46110>]))\n"
     ]
    }
   ],
   "source": [
    "for met in list(model.metabolites):\n",
    "    if met.id.startswith('MNXM'):\n",
    "        print(met.id, met.reactions)\n",
    "        model.remove_reactions(met.reactions)\n",
    "        if met.id in model.metabolites:\n",
    "            met.remove_from_model('destructive')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MNXR_rxns = [rxn for rxn in model.reactions if rxn.id.startswith('MNXR')]\n",
    "model.remove_reactions(MNXR_rxns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check for funktioning FBA after import:  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0 <= BIOMASS_REACTION_reverse_42726 <= 0, 0.1 <= BIOMASS_REACTION <= 1000.0}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.objective.variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "so4_in 0.0273340396411\n",
      "pi_in 0.0742559453479\n",
      "o2_in 27.6123082969\n",
      "co2_out 9.73776818209\n",
      "h2o_out 29.7680791941\n",
      "ACP_out 3.15502832193e-17\n",
      "hco3_out 2.37162203479\n",
      "h2o2_out 0.00108844802029\n",
      "co_out 4.66477722983e-06\n",
      "2dhp_out 0.045398233971\n",
      "5prdmbz_out 0.000155492574328\n",
      "EX_no3_e -1.3686798476\n",
      "ch4_in 18.405\n",
      "Fe_in 0.000951614554886\n",
      "BIOMASS_REACTION 0.155492574328\n"
     ]
    }
   ],
   "source": [
    "solution = show_uptake_excretion(model,model.reactions.get_by_id('BIOMASS_REACTION'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "model.remove_reactions([model.reactions.ATPM])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Demand and Sink reactions in the models are tagged either with an \\_out or \\_in suffix, while they should bear the prefix EX\\_. Transporters in the model have not been predicted properly. So far the only Transport-RXN that have been added denote simple diffusion of a few basic compounds. Transporter information will be inferred from Merlin's curated TCDB selection or from http://www.membranetransport.org/ output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EX_fdp_c: fdp_c --> \n",
      "EX_g3p_c: g3p_c --> \n",
      "EX_dhap_c: dhap_c --> \n",
      "EX_h_c: h_c --> \n",
      "EX_nadh_c: nadh_c --> \n",
      "EX_acald_c: acald_c --> \n",
      "EX_etoh_c: etoh_c --> \n",
      "EX_nad_c: nad_c --> \n",
      "EX_pyr_c: pyr_c --> \n",
      "EX_thmpp_c: thmpp_c --> \n",
      "EX_co2_c: co2_c --> \n",
      "EX_2ahethmpp_c: 2ahethmpp_c --> \n",
      "EX_dhlpro_c: dhlpro_c --> \n",
      "EX_accoa_c: accoa_c --> \n",
      "EX_coa_c: coa_c --> \n",
      "EX_atp_c: atp_c --> \n",
      "EX_adp_c: adp_c --> \n",
      "EX_pep_c: pep_c --> \n",
      "EX_h2o_c: h2o_c --> \n",
      "EX_2pg_c: 2pg_c --> \n",
      "EX_3pg_c: 3pg_c --> \n",
      "EX_13dpg_c: 13dpg_c --> \n",
      "EX_pi_c: pi_c --> \n",
      "EX_f6p_c: f6p_c --> \n",
      "EX_g6p_A_c: g6p_A_c --> \n",
      "EX_g1p_c: g1p_c --> \n",
      "EX_g6p_c: g6p_c --> \n",
      "EX_glc_bD_c: glc_bD_c --> \n",
      "EX_Glc_aD_c: Glc_aD_c --> \n",
      "EX_ppi_c: ppi_c --> \n",
      "EX_amp_c: amp_c --> \n",
      "EX_ac_c: ac_c --> \n",
      "EX_fum_c: fum_c --> \n",
      "EX_succ_c: succ_c --> \n",
      "EX_akg_c: akg_c --> \n",
      "EX_succoa_c: succoa_c --> \n",
      "EX_icit_c: icit_c --> \n",
      "EX_mal__L_c: mal__L_c --> \n",
      "EX_acon_C_c: acon_C_c --> \n",
      "EX_cit_c: cit_c --> \n",
      "EX_oaa_c: oaa_c --> \n",
      "EX_2ddg6p_c: 2ddg6p_c --> \n",
      "EX_prpp_c: prpp_c --> \n",
      "EX_r5p_c: r5p_c --> \n",
      "EX_s7p_c: s7p_c --> \n",
      "EX_xu5p_D_c: xu5p_D_c --> \n",
      "EX_ru5p_D_c: ru5p_D_c --> \n",
      "EX_e4p_c: e4p_c --> \n",
      "EX_nadph_c: nadph_c --> \n",
      "EX_6pgc_c: 6pgc_c --> \n",
      "EX_nadp_c: nadp_c --> \n",
      "EX_6pgl_c: 6pgl_c --> \n",
      "EX_r1p_c: r1p_c --> \n",
      "EX_actp_c: actp_c --> \n",
      "EX_fald_c: fald_c --> \n",
      "EX_ah6p__D_c: ah6p__D_c --> \n",
      "EX_udpg_c: udpg_c --> \n",
      "EX_utp_c: utp_c --> \n",
      "EX_udpglcur_c: udpglcur_c --> \n",
      "EX_man6p_c: man6p_c --> \n",
      "EX_man1p_c: man1p_c --> \n",
      "EX_gdpddman_c: gdpddman_c --> \n",
      "EX_gdpfuc_c: gdpfuc_c --> \n",
      "EX_gdpmann_c: gdpmann_c --> \n",
      "EX_gtp_c: gtp_c --> \n",
      "EX_glyald_c: glyald_c --> \n",
      "EX_f1p_c: f1p_c --> \n",
      "EX_tagdp_D_c: tagdp_D_c --> \n",
      "EX_tag6p_D_c: tag6p_D_c --> \n",
      "EX_cdp_c: cdp_c --> \n",
      "EX_ctp_c: ctp_c --> \n",
      "EX_udp_c: udp_c --> \n",
      "EX_idp_c: idp_c --> \n",
      "EX_itp_c: itp_c --> \n",
      "EX_sucr_c: sucr_c --> \n",
      "EX_fru_c: fru_c --> \n",
      "EX_glc__D_c: glc__D_c --> \n",
      "EX_udpgal_c: udpgal_c --> \n",
      "EX_gal_c: gal_c --> \n",
      "EX_toctd2eACP_c: toctd2eACP_c --> \n",
      "EX_ocdcaACP_c: ocdcaACP_c --> \n",
      "EX_3hoctaACP_c: 3hoctaACP_c --> \n",
      "EX_3ooctdACP_c: 3ooctdACP_c --> \n",
      "EX_ACP_c: ACP_c --> \n",
      "EX_malACP_c: malACP_c --> \n",
      "EX_palmACP_c: palmACP_c --> \n",
      "EX_acACP_c: acACP_c --> \n",
      "EX_actACP_c: actACP_c --> \n",
      "EX_3opalmACP_c: 3opalmACP_c --> \n",
      "EX_myrsACP_c: myrsACP_c --> \n",
      "EX_3omrsACP_c: 3omrsACP_c --> \n",
      "EX_ddcaACP_c: ddcaACP_c --> \n",
      "EX_3oddecACP_c: 3oddecACP_c --> \n",
      "EX_dcaACP_c: dcaACP_c --> \n",
      "EX_3odecACP_c: 3odecACP_c --> \n",
      "EX_ocACP_c: ocACP_c --> \n",
      "EX_3ooctACP_c: 3ooctACP_c --> \n",
      "EX_hexACP_c: hexACP_c --> \n",
      "EX_3ohexACP_c: 3ohexACP_c --> \n",
      "EX_butACP_c: butACP_c --> \n",
      "EX_tpalm2eACP_c: tpalm2eACP_c --> \n",
      "EX_3hpalmACP_c: 3hpalmACP_c --> \n",
      "EX_tmrs2eACP_c: tmrs2eACP_c --> \n",
      "EX_3hmrsACP_c: 3hmrsACP_c --> \n",
      "EX_tddec2eACP_c: tddec2eACP_c --> \n",
      "EX_3hddecACP_c: 3hddecACP_c --> \n",
      "EX_tdec2eACP_c: tdec2eACP_c --> \n",
      "EX_3hdecACP_c: 3hdecACP_c --> \n",
      "EX_toct2eACP_c: toct2eACP_c --> \n",
      "EX_3hoctACP_c: 3hoctACP_c --> \n",
      "EX_thex2eACP_c: thex2eACP_c --> \n",
      "EX_3hhexACP_c: 3hhexACP_c --> \n",
      "EX_but2eACP_c: but2eACP_c --> \n",
      "EX_3hbutACP_c: 3hbutACP_c --> \n",
      "EX_malcoa_c: malcoa_c --> \n",
      "EX_hco3_c: hco3_c --> \n",
      "EX_btnCCP_c: btnCCP_c --> \n",
      "EX_cbtnCCP_c: cbtnCCP_c --> \n",
      "EX_pmtcoa_c: pmtcoa_c --> \n",
      "EX_hdca_c: hdca_c --> \n",
      "EX_dsmsterol_c: dsmsterol_c --> \n",
      "EX_chsterol_c: chsterol_c --> \n",
      "EX_ddsmsterol_c: ddsmsterol_c --> \n",
      "EX_7dhchsterol_c: 7dhchsterol_c --> \n",
      "EX_chlstol_c: chlstol_c --> \n",
      "EX_lthstrl_c: lthstrl_c --> \n",
      "EX_zymst_c: zymst_c --> \n",
      "EX_zymstnl_c: zymstnl_c --> \n",
      "EX_44mzym_c: 44mzym_c --> \n",
      "EX_for_c: for_c --> \n",
      "EX_mergtrol_c: mergtrol_c --> \n",
      "EX_o2_c: o2_c --> \n",
      "EX_lanost_c: lanost_c --> \n",
      "EX_ahcys_c: ahcys_c --> \n",
      "EX_amet_c: amet_c --> \n",
      "EX_2omph_c: 2omph_c --> \n",
      "EX_2ohph_c: 2ohph_c --> \n",
      "EX_2oph_c: 2oph_c --> \n",
      "EX_3ophb_c: 3ophb_c --> \n",
      "EX_q8_im: q8_im --> \n",
      "EX_4hbz_c: 4hbz_c --> \n",
      "EX_chor_c: chor_c --> \n",
      "EX_octdp_c: octdp_c --> \n",
      "EX_phyQ_c: phyQ_c --> \n",
      "EX_dmtphllqne_c: dmtphllqne_c --> \n",
      "EX_2ombz_c: 2ombz_c --> \n",
      "EX_2ommb_c: 2ommb_c --> \n",
      "EX_trdrd_c: trdrd_c --> \n",
      "EX_gdp_c: gdp_c --> \n",
      "EX_trdox_c: trdox_c --> \n",
      "EX_dgdp_c: dgdp_c --> \n",
      "EX_gln__L_c: gln__L_c --> \n",
      "EX_pram_c: pram_c --> \n",
      "EX_glu__L_c: glu__L_c --> \n",
      "EX_gar_c: gar_c --> \n",
      "EX_gly_c: gly_c --> \n",
      "EX_fgam_c: fgam_c --> \n",
      "EX_thf_c: thf_c --> \n",
      "EX_10fthf_c: 10fthf_c --> \n",
      "EX_fpram_c: fpram_c --> \n",
      "EX_air_c: air_c --> \n",
      "EX_aicar_c: aicar_c --> \n",
      "EX_fprica_c: fprica_c --> \n",
      "EX_25aics_c: 25aics_c --> \n",
      "EX_asp__L_c: asp__L_c --> \n",
      "EX_5aizc_c: 5aizc_c --> \n",
      "EX_dna_c: dna_c --> \n",
      "EX_datp_c: datp_c --> \n",
      "EX_dadp_c: dadp_c --> \n",
      "EX_dcamp_c: dcamp_c --> \n",
      "EX_imp_c: imp_c --> \n",
      "EX_adn_c: adn_c --> \n",
      "EX_ade_c: ade_c --> \n",
      "EX_damp_c: damp_c --> \n",
      "EX_dad_2_c: dad_2_c --> \n",
      "EX_ins_c: ins_c --> \n",
      "EX_xmp_c: xmp_c --> \n",
      "EX_gmp_c: gmp_c --> \n",
      "EX_nh3_c: nh3_c --> \n",
      "EX_hxan_c: hxan_c --> \n",
      "EX_gsn_c: gsn_c --> \n",
      "EX_xan_c: xan_c --> \n",
      "EX_xtsn_c: xtsn_c --> \n",
      "EX_gua_c: gua_c --> \n",
      "EX_ppgpp_c: ppgpp_c --> \n",
      "EX_gdptp_c: gdptp_c --> \n",
      "EX_dgmp_c: dgmp_c --> \n",
      "EX_dgtp_c: dgtp_c --> \n",
      "EX_dgsn_c: dgsn_c --> \n",
      "EX_ap4a_c: ap4a_c --> \n",
      "EX_so4_c: so4_c --> \n",
      "EX_aps_c: aps_c --> \n",
      "EX_paps_c: paps_c --> \n",
      "EX_xtp_c: xtp_c --> \n",
      "EX_adprib_c: adprib_c --> \n",
      "EX_didp_c: didp_c --> \n",
      "EX_ditp_c: ditp_c --> \n",
      "EX_dimp_c: dimp_c --> \n",
      "EX_C04051_c: C04051_c --> \n",
      "EX_5caiz_c: 5caiz_c --> \n",
      "EX_dudp_c: dudp_c --> \n",
      "EX_ump_c: ump_c --> \n",
      "EX_cbp_c: cbp_c --> \n",
      "EX_cbasp_c: cbasp_c --> \n",
      "EX_dhor_S_c: dhor_S_c --> \n",
      "EX_orot_c: orot_c --> \n",
      "EX_orot5p_c: orot5p_c --> \n",
      "EX_ura_c: ura_c --> \n",
      "EX_uri_c: uri_c --> \n",
      "EX_cmp_c: cmp_c --> \n",
      "EX_cytd_c: cytd_c --> \n",
      "EX_csn_c: csn_c --> \n",
      "EX_dcdp_c: dcdp_c --> \n",
      "EX_dctp_c: dctp_c --> \n",
      "EX_dutp_c: dutp_c --> \n",
      "EX_dcmp_c: dcmp_c --> \n",
      "EX_dcyt_c: dcyt_c --> \n",
      "EX_dump_c: dump_c --> \n",
      "EX_duri_c: duri_c --> \n",
      "EX_2dr1p_c: 2dr1p_c --> \n",
      "EX_mlthf_c: mlthf_c --> \n",
      "EX_dtmp_c: dtmp_c --> \n",
      "EX_dhf_c: dhf_c --> \n",
      "EX_dttp_c: dttp_c --> \n",
      "EX_dtdp_c: dtdp_c --> \n",
      "EX_thymd_c: thymd_c --> \n",
      "EX_thym_c: thym_c --> \n",
      "EX_5mcsn_c: 5mcsn_c --> \n",
      "EX_asn__L_c: asn__L_c --> \n",
      "EX_h2o2_c: h2o2_c --> \n",
      "EX_argsuc_c: argsuc_c --> \n",
      "EX_citr__L_c: citr__L_c --> \n",
      "EX_arg__L_c: arg__L_c --> \n",
      "EX_ala__L_c: ala__L_c --> \n",
      "EX_glx_c: glx_c --> \n",
      "EX_sucsal_c: sucsal_c --> \n",
      "EX_gam6p_c: gam6p_c --> \n",
      "EX_glyc_R_c: glyc_R_c --> \n",
      "EX_ser__L_c: ser__L_c --> \n",
      "EX_hpyr_c: hpyr_c --> \n",
      "EX_4pasp_c: 4pasp_c --> \n",
      "EX_pser__L_c: pser__L_c --> \n",
      "EX_3php_c: 3php_c --> \n",
      "EX_2obut_c: 2obut_c --> \n",
      "EX_thr__L_c: thr__L_c --> \n",
      "EX_phom_c: phom_c --> \n",
      "EX_aspsa_c: aspsa_c --> \n",
      "EX_alpro_c: alpro_c --> \n",
      "EX_lpro_c: lpro_c --> \n",
      "EX_hom__L_c: hom__L_c --> \n",
      "EX_trp__L_c: trp__L_c --> \n",
      "EX_3ig3p_c: 3ig3p_c --> \n",
      "EX_suchms_c: suchms_c --> \n",
      "EX_h2s_c: h2s_c --> \n",
      "EX_hcys__L_c: hcys__L_c --> \n",
      "EX_2kmb_c: 2kmb_c --> \n",
      "EX_12dmpo_c: 12dmpo_c --> \n",
      "EX_3mtpa_c: 3mtpa_c --> \n",
      "EX_co_c: co_c --> \n",
      "EX_2h3kmtp_c: 2h3kmtp_c --> \n",
      "EX_5mdru1p_c: 5mdru1p_c --> \n",
      "EX_5mdr1p_c: 5mdr1p_c --> \n",
      "EX_5mta_c: 5mta_c --> \n",
      "EX_5mtr_c: 5mtr_c --> \n",
      "EX_spmd_c: spmd_c --> \n",
      "EX_ptrc_c: ptrc_c --> \n",
      "EX_ametam_c: ametam_c --> \n",
      "EX_met__L_c: met__L_c --> \n",
      "EX_5mthf_c: 5mthf_c --> \n",
      "EX_cyst__L_c: cyst__L_c --> \n",
      "EX_5mthglu_c: 5mthglu_c --> \n",
      "EX_hpglu_c: hpglu_c --> \n",
      "EX_dna5mtc_c: dna5mtc_c --> \n",
      "EX_dnac_c: dnac_c --> \n",
      "EX_rhcys_c: rhcys_c --> \n",
      "EX_cys__L_c: cys__L_c --> \n",
      "EX_acser_c: acser_c --> \n",
      "EX_tsul_c: tsul_c --> \n",
      "EX_so3_c: so3_c --> \n",
      "EX_scys__L_c: scys__L_c --> \n",
      "EX_Lcyst_c: Lcyst_c --> \n",
      "EX_3spyr_c: 3spyr_c --> \n",
      "EX_Lcystin_c: Lcystin_c --> \n",
      "EX_thcys_c: thcys_c --> \n",
      "EX_3sala_c: 3sala_c --> \n",
      "EX_3snpyr_c: 3snpyr_c --> \n",
      "EX_mercppyr_c: mercppyr_c --> \n",
      "EX_3mob_c: 3mob_c --> \n",
      "EX_val__L_c: val__L_c --> \n",
      "EX_ile__L_c: ile__L_c --> \n",
      "EX_3mop_c: 3mop_c --> \n",
      "EX_4mop_c: 4mop_c --> \n",
      "EX_leu__L_c: leu__L_c --> \n",
      "EX_citac_c: citac_c --> \n",
      "EX_r3mmal_c: r3mmal_c --> \n",
      "EX_3hmop_c: 3hmop_c --> \n",
      "EX_23dhmp_c: 23dhmp_c --> \n",
      "EX_3c3hmp_c: 3c3hmp_c --> \n",
      "EX_2ippm_c: 2ippm_c --> \n",
      "EX_3c4mop_c: 3c4mop_c --> \n",
      "EX_3c2hmp_c: 3c2hmp_c --> \n",
      "EX_23dhmb_c: 23dhmb_c --> \n",
      "EX_2ahbut_c: 2ahbut_c --> \n",
      "EX_3hmoa_c: 3hmoa_c --> \n",
      "EX_alac_S_c: alac_S_c --> \n",
      "EX_ugmda_c: ugmda_c --> \n",
      "EX_ugmd_c: ugmd_c --> \n",
      "EX_alaala_c: alaala_c --> \n",
      "EX_23dhdp_c: 23dhdp_c --> \n",
      "EX_thdp_c: thdp_c --> \n",
      "EX_sl2a6o_c: sl2a6o_c --> \n",
      "EX_sl26da_c: sl26da_c --> \n",
      "EX_26dap_LL_c: 26dap_LL_c --> \n",
      "EX_26dap_M_c: 26dap_M_c --> \n",
      "EX_lys__L_c: lys__L_c --> \n",
      "EX_hcit_c: hcit_c --> \n",
      "EX_uamag_c: uamag_c --> \n",
      "EX_lys_D_c: lys_D_c --> \n",
      "EX_6a2ohxnt_c: 6a2ohxnt_c --> \n",
      "EX_2oxoadp_c: 2oxoadp_c --> \n",
      "EX_S_gtrdhdlp_c: S_gtrdhdlp_c --> \n",
      "EX_glutcoa_c: glutcoa_c --> \n",
      "EX_6apa_c: 6apa_c --> \n",
      "EX_acglu_c: acglu_c --> \n",
      "EX_acg5p_c: acg5p_c --> \n",
      "EX_acg5sa_c: acg5sa_c --> \n",
      "EX_acorn_c: acorn_c --> \n",
      "EX_orn_c: orn_c --> \n",
      "EX_glu5sa_c: glu5sa_c --> \n",
      "EX_1pyr5c_c: 1pyr5c_c --> \n",
      "EX_pro__L_c: pro__L_c --> \n",
      "EX_pepd_c: pepd_c --> \n",
      "EX_1p3h5c_c: 1p3h5c_c --> \n",
      "EX_4hpro_LT_c: 4hpro_LT_c --> \n",
      "EX_e4hglu_c: e4hglu_c --> \n",
      "EX_asp_D_c: asp_D_c --> \n",
      "EX_agm_c: agm_c --> \n",
      "EX_Ncbmpts_c: Ncbmpts_c --> \n",
      "EX_glu5p_c: glu5p_c --> \n",
      "EX_crtn_c: crtn_c --> \n",
      "EX_cspmd_c: cspmd_c --> \n",
      "EX_sprm_c: sprm_c --> \n",
      "EX_his__L_c: his__L_c --> \n",
      "EX_hista_c: hista_c --> \n",
      "EX_prbatp_c: prbatp_c --> \n",
      "EX_histd_c: histd_c --> \n",
      "EX_imacp_c: imacp_c --> \n",
      "EX_hisp_c: hisp_c --> \n",
      "EX_eig3p_c: eig3p_c --> \n",
      "EX_prfp_c: prfp_c --> \n",
      "EX_prlp_c: prlp_c --> \n",
      "EX_prbamp_c: prbamp_c --> \n",
      "EX_34dhoxpeg_c: 34dhoxpeg_c --> \n",
      "EX_34dhmald_c: 34dhmald_c --> \n",
      "EX_tyr__L_c: tyr__L_c --> \n",
      "EX_34hpp_c: 34hpp_c --> \n",
      "EX_phaccoa_c: phaccoa_c --> \n",
      "EX_pac_c: pac_c --> \n",
      "EX_phe__L_c: phe__L_c --> \n",
      "EX_phpyr_c: phpyr_c --> \n",
      "EX_glu_D_c: glu_D_c --> \n",
      "EX_cinnm_c: cinnm_c --> \n",
      "EX_cenchddd_c: cenchddd_c --> \n",
      "EX_pppn_c: pppn_c --> \n",
      "EX_cechddd_c: cechddd_c --> \n",
      "EX_indpyr_c: indpyr_c --> \n",
      "EX_id3acald_c: id3acald_c --> \n",
      "EX_ind3acnl_c: ind3acnl_c --> \n",
      "EX_ind3ac_c: ind3ac_c --> \n",
      "EX_2dda7p_c: 2dda7p_c --> \n",
      "EX_3psme_c: 3psme_c --> \n",
      "EX_indole_c: indole_c --> \n",
      "EX_2cpr5p_c: 2cpr5p_c --> \n",
      "EX_pran_c: pran_c --> \n",
      "EX_anth_c: anth_c --> \n",
      "EX_pphn_c: pphn_c --> \n",
      "EX_Largn_c: Largn_c --> \n",
      "EX_skm5p_c: skm5p_c --> \n",
      "EX_skm_c: skm_c --> \n",
      "EX_3dhsk_c: 3dhsk_c --> \n",
      "EX_3dhq_c: 3dhq_c --> \n",
      "EX_pnto_R_c: pnto_R_c --> \n",
      "EX_ala_B_c: ala_B_c --> \n",
      "EX_pant_R_c: pant_R_c --> \n",
      "EX_taur_c: taur_c --> \n",
      "EX_seln_c: seln_c --> \n",
      "EX_selt_c: selt_c --> \n",
      "EX_selmeth_c: selmeth_c --> \n",
      "EX_trnamet_c: trnamet_c --> \n",
      "EX_selcys_c: selcys_c --> \n",
      "EX_sel_c: sel_c --> \n",
      "EX_adsel_c: adsel_c --> \n",
      "EX_selhcys_c: selhcys_c --> \n",
      "EX_selcyst_c: selcyst_c --> \n",
      "EX_aprop_c: aprop_c --> \n",
      "EX_acybut_c: acybut_c --> \n",
      "EX_uama_c: uama_c --> \n",
      "EX_uamr_c: uamr_c --> \n",
      "EX_orn_D_c: orn_D_c --> \n",
      "EX_5a2opntn_c: 5a2opntn_c --> \n",
      "EX_arg_D_c: arg_D_c --> \n",
      "EX_5g2oxpt_c: 5g2oxpt_c --> \n",
      "EX_ala_D_c: ala_D_c --> \n",
      "EX_gthrd_c: gthrd_c --> \n",
      "EX_gthox_c: gthox_c --> \n",
      "EX_glucys_c: glucys_c --> \n",
      "EX_cgly_c: cgly_c --> \n",
      "EX_na15dap_c: na15dap_c --> \n",
      "EX_15dap_c: 15dap_c --> \n",
      "EX_cellb_c: cellb_c --> \n",
      "EX_udpgalur_c: udpgalur_c --> \n",
      "EX_adpglc_c: adpglc_c --> \n",
      "EX_malt_c: malt_c --> \n",
      "EX_tre_c: tre_c --> \n",
      "EX_uacgam_c: uacgam_c --> \n",
      "EX_acgam1p_c: acgam1p_c --> \n",
      "EX_gam1p_c: gam1p_c --> \n",
      "EX_uaccg_c: uaccg_c --> \n",
      "EX_chtbs_c: chtbs_c --> \n",
      "EX_acgam_c: acgam_c --> \n",
      "EX_mi3p_D_c: mi3p_D_c --> \n",
      "EX_inost_c: inost_c --> \n",
      "EX_dtdp4d6dm_c: dtdp4d6dm_c --> \n",
      "EX_dtdprmn_c: dtdprmn_c --> \n",
      "EX_dtdp4d6dg_c: dtdp4d6dg_c --> \n",
      "EX_dtdpglu_c: dtdpglu_c --> \n",
      "EX_ckdo_c: ckdo_c --> \n",
      "EX_kdo_c: kdo_c --> \n",
      "EX_kdo8p_c: kdo8p_c --> \n",
      "EX_lipidA_c: lipidA_c --> \n",
      "EX_lipidAds_c: lipidAds_c --> \n",
      "EX_u23ga_c: u23ga_c --> \n",
      "EX_lipidX_c: lipidX_c --> \n",
      "EX_gmhep7p_c: gmhep7p_c --> \n",
      "EX_gmhep17bp_c: gmhep17bp_c --> \n",
      "EX_gmhep1p_c: gmhep1p_c --> \n",
      "EX_adphep_DD_c: adphep_DD_c --> \n",
      "EX_kdo2lipid4_c: kdo2lipid4_c --> \n",
      "EX_kdolipid4_c: kdolipid4_c --> \n",
      "EX_ara5p_c: ara5p_c --> \n",
      "EX_u3hga_c: u3hga_c --> \n",
      "EX_u3aga_c: u3aga_c --> \n",
      "EX_udcpdp_c: udcpdp_c --> \n",
      "EX_uaaGgla_c: uaaGgla_c --> \n",
      "EX_uaGgla_c: uaGgla_c --> \n",
      "EX_uGgla_c: uGgla_c --> \n",
      "EX_uGgl_c: uGgl_c --> \n",
      "EX_uagmda_c: uagmda_c --> \n",
      "EX_uaagmda_c: uaagmda_c --> \n",
      "EX_acoa_c: acoa_c --> \n",
      "EX_glyc1p_c: glyc1p_c --> \n",
      "EX_dha_c: dha_c --> \n",
      "EX_12dgr_EC_c: 12dgr_EC_c --> \n",
      "EX_mi4p_D_c: mi4p_D_c --> \n",
      "EX_mi1p__D_c: mi1p__D_c --> \n",
      "EX_g3pe_c: g3pe_c --> \n",
      "EX_etha_c: etha_c --> \n",
      "EX_pgp_EC_c: pgp_EC_c --> \n",
      "EX_ptd2meeta_SC_c: ptd2meeta_SC_c --> \n",
      "EX_g3pc_c: g3pc_c --> \n",
      "EX_chol_c: chol_c --> \n",
      "EX_15HPET_c: 15HPET_c --> \n",
      "EX_5HPET_c: 5HPET_c --> \n",
      "EX_leuktrC4_c: leuktrC4_c --> \n",
      "EX_leuktrD4_c: leuktrD4_c --> \n",
      "EX_mthgxl_c: mthgxl_c --> \n",
      "EX_lgt_S_c: lgt_S_c --> \n",
      "EX_lac_D_c: lac_D_c --> \n",
      "EX_tol_c: tol_c --> \n",
      "EX_4crsol_c: 4crsol_c --> \n",
      "EX_fdxrd_c: fdxrd_c --> \n",
      "EX_fdxox_c: fdxox_c --> \n",
      "EX_bzalc_c: bzalc_c --> \n",
      "EX_cl_c: cl_c --> \n",
      "EX_4nph_c: 4nph_c --> \n",
      "EX_24nph_c: 24nph_c --> \n",
      "EX_no2_c: no2_c --> \n",
      "EX_catechol_c: catechol_c --> \n",
      "EX_phen_c: phen_c --> \n",
      "EX_bz_c: bz_c --> \n",
      "EX_4h2oglt_c: 4h2oglt_c --> \n",
      "EX_glyclt_c: glyclt_c --> \n",
      "EX_2pglyc_c: 2pglyc_c --> \n",
      "EX_rb15bp_c: rb15bp_c --> \n",
      "EX_micit_c: micit_c --> \n",
      "EX_2mcacn_c: 2mcacn_c --> \n",
      "EX_ppcoa_c: ppcoa_c --> \n",
      "EX_ppa_c: ppa_c --> \n",
      "EX_ppap_c: ppap_c --> \n",
      "EX_fad_c: fad_c --> \n",
      "EX_fadh2_c: fadh2_c --> \n",
      "EX_actn_R_c: actn_R_c --> \n",
      "EX_itacon_c: itacon_c --> \n",
      "EX_itaccoa_c: itaccoa_c --> \n",
      "EX_methf_c: methf_c --> \n",
      "EX_fol_c: fol_c --> \n",
      "EX_5fthf_c: 5fthf_c --> \n",
      "EX_fmettrna_c: fmettrna_c --> \n",
      "EX_mettrna_c: mettrna_c --> \n",
      "EX_h2_c: h2_c --> \n",
      "EX_focytC_c: focytC_c --> \n",
      "EX_meoh_c: meoh_c --> \n",
      "EX_ficytC_c: ficytC_c --> \n",
      "EX_510mh4mpt_c: 510mh4mpt_c --> \n",
      "EX_h4mpt_c: h4mpt_c --> \n",
      "EX_510methmpt_c: 510methmpt_c --> \n",
      "EX_5fthmpt_c: 5fthmpt_c --> \n",
      "EX_formmfr_c: formmfr_c --> \n",
      "EX_mfr_c: mfr_c --> \n",
      "EX_ch4_c: ch4_c --> \n",
      "EX_s17bp_c: s17bp_c --> \n",
      "EX_thmmp_c: thmmp_c --> \n",
      "EX_dhgly_c: dhgly_c --> \n",
      "EX_4mpetz_c: 4mpetz_c --> \n",
      "EX_dxyl5p_c: dxyl5p_c --> \n",
      "EX_4ahmmp_c: 4ahmmp_c --> \n",
      "EX_2mahmp_c: 2mahmp_c --> \n",
      "EX_4ampm_c: 4ampm_c --> \n",
      "EX_db4p_c: db4p_c --> \n",
      "EX_fmn_c: fmn_c --> \n",
      "EX_ribflv_c: ribflv_c --> \n",
      "EX_25dhpp_c: 25dhpp_c --> \n",
      "EX_5apru_c: 5apru_c --> \n",
      "EX_5aprbu_c: 5aprbu_c --> \n",
      "EX_dmlz_c: dmlz_c --> \n",
      "EX_4r5au_c: 4r5au_c --> \n",
      "EX_dmbzid_c: dmbzid_c --> \n",
      "EX_fmnh2_c: fmnh2_c --> \n",
      "EX_phthr_c: phthr_c --> \n",
      "EX_4hthr_c: 4hthr_c --> \n",
      "EX_ohpb_c: ohpb_c --> \n",
      "EX_pdx5p_c: pdx5p_c --> \n",
      "EX_iasp_c: iasp_c --> \n",
      "EX_quln_c: quln_c --> \n",
      "EX_nicrnt_c: nicrnt_c --> \n",
      "EX_nicrns_c: nicrns_c --> \n",
      "EX_dnad_c: dnad_c --> \n",
      "EX_nac_c: nac_c --> \n",
      "EX_ncam_c: ncam_c --> \n",
      "EX_nmn_c: nmn_c --> \n",
      "EX_rnam_c: rnam_c --> \n",
      "EX_dpcoa_c: dpcoa_c --> \n",
      "EX_pan4p_c: pan4p_c --> \n",
      "EX_ptth_c: ptth_c --> \n",
      "EX_4ppcys_c: 4ppcys_c --> \n",
      "EX_ptcys_c: ptcys_c --> \n",
      "EX_4ppan_c: 4ppan_c --> \n",
      "EX_2dhp_c: 2dhp_c --> \n",
      "EX_apoACP_c: apoACP_c --> \n",
      "EX_pap_c: pap_c --> \n",
      "EX_btamp_c: btamp_c --> \n",
      "EX_btn_c: btn_c --> \n",
      "EX_dtbt_c: dtbt_c --> \n",
      "EX_dad_5_c: dad_5_c --> \n",
      "EX_dann_c: dann_c --> \n",
      "EX_8aonn_c: 8aonn_c --> \n",
      "EX_amob_c: amob_c --> \n",
      "EX_pmcoa_c: pmcoa_c --> \n",
      "EX_pimACP_c: pimACP_c --> \n",
      "EX_pmeACP_c: pmeACP_c --> \n",
      "EX_hgmeACP_c: hgmeACP_c --> \n",
      "EX_egmeACP_c: egmeACP_c --> \n",
      "EX_gmeACP_c: gmeACP_c --> \n",
      "EX_hpmeACP_c: hpmeACP_c --> \n",
      "EX_epmeACP_c: epmeACP_c --> \n",
      "EX_4adcho_c: 4adcho_c --> \n",
      "EX_4abz_c: 4abz_c --> \n",
      "EX_thfglu_c: thfglu_c --> \n",
      "EX_dhpt_c: dhpt_c --> \n",
      "EX_2ahhmd_c: 2ahhmd_c --> \n",
      "EX_2ahhmp_c: 2ahhmp_c --> \n",
      "EX_dhnpt_c: dhnpt_c --> \n",
      "EX_gcald_c: gcald_c --> \n",
      "EX_ahdt_c: ahdt_c --> \n",
      "EX_ptp_c: ptp_c --> \n",
      "EX_pppi_c: pppi_c --> \n",
      "EX_6pthp_c: 6pthp_c --> \n",
      "EX_cpmp_c: cpmp_c --> \n",
      "EX_6ctd_c: 6ctd_c --> \n",
      "EX_cdg_c: cdg_c --> \n",
      "EX_preq0_c: preq0_c --> \n",
      "EX_preq1_c: preq1_c --> \n",
      "EX_pppg9_c: pppg9_c --> \n",
      "EX_cpppg3_c: cpppg3_c --> \n",
      "EX_pheme_c: pheme_c --> \n",
      "EX_frdp_c: frdp_c --> \n",
      "EX_hemeO_c: hemeO_c --> \n",
      "EX_fe2_c: fe2_c --> \n",
      "EX_ppp9_c: ppp9_c --> \n",
      "EX_uppg3_c: uppg3_c --> \n",
      "EX_hmbil_c: hmbil_c --> \n",
      "EX_dscl_c: dscl_c --> \n",
      "EX_uppg1_c: uppg1_c --> \n",
      "EX_cpppg1_c: cpppg1_c --> \n",
      "EX_ppbng_c: ppbng_c --> \n",
      "EX_5aop_c: 5aop_c --> \n",
      "EX_glu1sa_c: glu1sa_c --> \n",
      "EX_glutrna_c: glutrna_c --> \n",
      "EX_trnaglu_c: trnaglu_c --> \n",
      "EX_sheme_c: sheme_c --> \n",
      "EX_scl_c: scl_c --> \n",
      "EX_copre2_c: copre2_c --> \n",
      "EX_copre3_c: copre3_c --> \n",
      "EX_copre4_c: copre4_c --> \n",
      "EX_copre5_c: copre5_c --> \n",
      "EX_pre3a_c: pre3a_c --> \n",
      "EX_pre4_c: pre4_c --> \n",
      "EX_pre3b_c: pre3b_c --> \n",
      "EX_pre5_c: pre5_c --> \n",
      "EX_pre8_c: pre8_c --> \n",
      "EX_hgbyr_c: hgbyr_c --> \n",
      "EX_hgbam_c: hgbam_c --> \n",
      "EX_cobya_c: cobya_c --> \n",
      "EX_copre8_c: copre8_c --> \n",
      "EX_co2dam_c: co2dam_c --> \n",
      "EX_co1dam_c: co1dam_c --> \n",
      "EX_adcobdam_c: adcobdam_c --> \n",
      "EX_adcobhex_c: adcobhex_c --> \n",
      "EX_cbi_c: cbi_c --> \n",
      "EX_adocbi_c: adocbi_c --> \n",
      "EX_adocbip_c: adocbip_c --> \n",
      "EX_agdpcbi_c: agdpcbi_c --> \n",
      "EX_adocbl_c: adocbl_c --> \n",
      "EX_rdmbzi_c: rdmbzi_c --> \n",
      "EX_5prdmbz_c: 5prdmbz_c --> \n",
      "EX_hemeA__1_c: hemeA__1_c --> \n",
      "EX_appl_c: appl_c --> \n",
      "EX_applp_c: applp_c --> \n",
      "EX_pre6b_c: pre6b_c --> \n",
      "EX_copre6_c: copre6_c --> \n",
      "EX_codscl5b_c: codscl5b_c --> \n",
      "EX_2me4p_c: 2me4p_c --> \n",
      "EX_4c2me_c: 4c2me_c --> \n",
      "EX_2p4c2me_c: 2p4c2me_c --> \n",
      "EX_2mecdp_c: 2mecdp_c --> \n",
      "EX_h2mb4p_c: h2mb4p_c --> \n",
      "EX_dmpp_c: dmpp_c --> \n",
      "EX_ipdp_c: ipdp_c --> \n",
      "EX_grdp_c: grdp_c --> \n",
      "EX_psqldp_c: psqldp_c --> \n",
      "EX_sql_c: sql_c --> \n",
      "EX_dplpt_c: dplpt_c --> \n",
      "EX_dpterol_c: dpterol_c --> \n",
      "EX_ham_c: ham_c --> \n",
      "EX_n2_c: n2_c --> \n",
      "EX_n2o_c: n2o_c --> \n",
      "EX_no_c: no_c --> \n",
      "EX_flxr_c: flxr_c --> \n",
      "EX_flxso_c: flxso_c --> \n",
      "EX_iletrna_c: iletrna_c --> \n",
      "EX_trnaile_c: trnaile_c --> \n",
      "EX_histrna_c: histrna_c --> \n",
      "EX_trnahis_c: trnahis_c --> \n",
      "EX_phetrna_c: phetrna_c --> \n",
      "EX_trnaphe_c: trnaphe_c --> \n",
      "EX_argtrna_c: argtrna_c --> \n",
      "EX_trnaarg_c: trnaarg_c --> \n",
      "EX_trnacys_c: trnacys_c --> \n",
      "EX_cystrna_c: cystrna_c --> \n",
      "EX_glytrna_c: glytrna_c --> \n",
      "EX_trnagly_c: trnagly_c --> \n",
      "EX_trnaasp_c: trnaasp_c --> \n",
      "EX_asptrna_c: asptrna_c --> \n",
      "EX_sertrna_c: sertrna_c --> \n",
      "EX_trnaser_c: trnaser_c --> \n",
      "EX_valtrna_c: valtrna_c --> \n",
      "EX_trnaval_c: trnaval_c --> \n",
      "EX_alatrna_c: alatrna_c --> \n",
      "EX_trnaala_c: trnaala_c --> \n",
      "EX_lystrna_c: lystrna_c --> \n",
      "EX_leutrna_c: leutrna_c --> \n",
      "EX_trnaleu_c: trnaleu_c --> \n",
      "EX_thrtrna_c: thrtrna_c --> \n",
      "EX_trnathr_c: trnathr_c --> \n",
      "EX_trptrna_c: trptrna_c --> \n",
      "EX_trnatrp_c: trnatrp_c --> \n",
      "EX_tyrtrna_c: tyrtrna_c --> \n",
      "EX_trnatyr_c: trnatyr_c --> \n",
      "EX_protrna_c: protrna_c --> \n",
      "EX_trnapro_c: trnapro_c --> \n",
      "EX_glntrna_c: glntrna_c --> \n",
      "EX_asntrna_c: asntrna_c --> \n",
      "EX_sertrna_sec__c: sertrna_sec__c --> \n",
      "EX_odecoa_c: odecoa_c --> \n",
      "EX_ocdcea_c: ocdcea_c --> \n",
      "EX_stcoa_c: stcoa_c --> \n",
      "EX_ocdca_c: ocdca_c --> \n",
      "EX_3ohodcoa_c: 3ohodcoa_c --> \n",
      "EX_3hodcoa_c: 3hodcoa_c --> \n",
      "EX_focytb5_c: focytb5_c --> \n",
      "EX_ficytb5_c: ficytb5_c --> \n",
      "EX_octeACP_c: octeACP_c --> \n",
      "EX_glyb_c: glyb_c --> \n",
      "EX_dmgly_c: dmgly_c --> \n",
      "EX_pqqh2_c: pqqh2_c --> \n",
      "EX_pqq_c: pqq_c --> \n",
      "EX_trnalys_c: trnalys_c --> \n",
      "EX_dolp_c: dolp_c --> \n",
      "EX_mser__L_c: mser__L_c --> \n",
      "EX_retinal_c: retinal_c --> \n",
      "EX_retinol_c: retinol_c --> \n",
      "EX_f420_2h2_c: f420_2h2_c --> \n",
      "EX_f420_2_c: f420_2_c --> \n",
      "EX_q8h2_im: q8h2_im --> \n",
      "EX_2mhob_c: 2mhob_c --> \n",
      "EX_2mbcoa_c: 2mbcoa_c --> \n",
      "EX_bdg2hc_c: bdg2hc_c --> \n",
      "EX_2coum_c: 2coum_c --> \n",
      "EX_citmcoa__L_c: citmcoa__L_c --> \n",
      "EX_na1_c: na1_c --> \n",
      "EX_k_c: k_c --> \n",
      "EX_mg2_c: mg2_c --> \n",
      "EX_ca2_c: ca2_c --> \n",
      "EX_ttdca_c_None_: ttdca_c_None_ --> \n",
      "EX_ptdca_c_None_: ptdca_c_None_ --> \n",
      "EX_hpdca_c_None_: hpdca_c_None_ --> \n",
      "EX_ttdcea_c_None_: ttdcea_c_None_ --> \n",
      "EX_cpoa2h_c_None_: cpoa2h_c_None_ --> \n"
     ]
    }
   ],
   "source": [
    "demands = [rxn for rxn in model.reactions if rxn.id.endswith('_out')]\n",
    "\n",
    "for rxn in demands:\n",
    "    if rxn.reactants != []:\n",
    "        new_id = rxn.reactants[0].id\n",
    "        descriptive_name = rxn.reactants[0].name\n",
    "                \n",
    "        try: \n",
    "            model.reactions.get_by_id('EX_' + new_id)\n",
    "        except KeyError:\n",
    "            rxn.name = '{} exchange'.format(descriptive_name)\n",
    "            rxn.id = 'EX_' + new_id\n",
    "            print rxn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.1554925743277103"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.solve().f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ch4_in\n",
      "glycogen_in\n",
      "Fe_in\n"
     ]
    }
   ],
   "source": [
    "exchanges = [rxn for rxn in model.reactions if rxn.id.endswith('_in')]\n",
    "make_reversible = []\n",
    "for rxn in exchanges:\n",
    "    if rxn.products != []:\n",
    "        new_id = rxn.products[0].id\n",
    "        descriptive_name = rxn.products[0].name\n",
    "                \n",
    "        try: \n",
    "            model.reactions.get_by_id('EX_' + new_id)\n",
    "            make_reversible.append(rxn)\n",
    "        except KeyError:\n",
    "            print rxn.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rxn = model.reactions.Fe_in\n",
    "new_id = rxn.products[0].id\n",
    "descriptive_name = rxn.products[0].name\n",
    "rxn.id = 'EX_' + new_id\n",
    "rxn.name = '{} exchange'.format(descriptive_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rxn = model.reactions.glycogen_in\n",
    "new_id = rxn.products[0].id\n",
    "descriptive_name = rxn.products[0].name\n",
    "rxn.id = 'EX_' + new_id\n",
    "rxn.name = '{} exchange'.format(descriptive_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rxn = model.reactions.ch4_in\n",
    "new_id = rxn.products[0].id\n",
    "descriptive_name = rxn.products[0].name\n",
    "rxn.id = 'EX_' + new_id\n",
    "rxn.name = '{} exchange'.format(descriptive_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.1554925743277103"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.solve().f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Reaction nh3_in at 0x10db46610>,\n",
       " <Reaction na1_in at 0x10db46690>,\n",
       " <Reaction k_in at 0x10db46710>,\n",
       " <Reaction mg2_in at 0x10db46790>,\n",
       " <Reaction ca2_in at 0x10db46810>,\n",
       " <Reaction so4_in at 0x10db46890>,\n",
       " <Reaction cl_in at 0x10db46910>,\n",
       " <Reaction pi_in at 0x10db46990>,\n",
       " <Reaction h_in at 0x10db46a10>,\n",
       " <Reaction h2o_in at 0x10db46a90>,\n",
       " <Reaction co2_in at 0x10db46b10>,\n",
       " <Reaction o2_in at 0x10db46b90>]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.remove_reactions(make_reversible[0:2])\n",
    "make_reversible[2:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nh3_in 0.155492574328\n",
      "na1_in 0.22201063896\n",
      "k_in 0.22201063896\n",
      "mg2_in 0.22201063896\n",
      "ca2_in 0.22201063896\n",
      "so4_in\n",
      "cl_in 0.22201063896\n",
      "pi_in\n",
      "h_in 0.22201063896\n",
      "h2o_in 0.22201063896\n",
      "co2_in 0.22201063896\n",
      "o2_in\n"
     ]
    }
   ],
   "source": [
    "for rxn in make_reversible[2:]:\n",
    "    rxn.lower_bound = 0\n",
    "    rxn.upper_bound = 0\n",
    "    try:\n",
    "        solution = model.solve()\n",
    "        print rxn.id,solution.f\n",
    "    except:\n",
    "        print rxn.id\n",
    "    rxn.lower_bound = -1000\n",
    "    rxn.upper_bound = 1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "model.reactions.EX_so4_c.lower_bound = -1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "model.reactions.EX_pi_c.lower_bound = -1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "model.reactions.EX_o2_c.lower_bound = -1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "model.remove_reactions(make_reversible[2:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.1554925743277102"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.solve().f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EX_co2_c 9.80917347274\n",
      "EX_h2o_c 29.839484494\n",
      "EX_pi_c -0.0742559453479\n",
      "EX_ACP_c 2.23345647532e-17\n",
      "EX_hco3_c 2.30021674414\n",
      "EX_o2_c -27.6123082969\n",
      "EX_trdox_c 4.62223186653e-33\n",
      "EX_so4_c -0.0273340396411\n",
      "EX_h2o2_c 0.00108844802029\n",
      "EX_co_c 4.66477722983e-06\n",
      "EX_2dhp_c 0.045398233971\n",
      "EX_dhpt_c 8.67361737989e-19\n",
      "EX_5prdmbz_c 0.000155492574328\n",
      "EX_no3_e -1.3686798476\n",
      "EX_ch4_p 18.405\n",
      "EX_fe2_p_None_ 0.000951614554886\n",
      "BIOMASS_REACTION 0.155492574328\n"
     ]
    }
   ],
   "source": [
    "solution = show_uptake_excretion(model,model.reactions.get_by_id('BIOMASS_REACTION'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "model.remove_reactions([rxn for rxn in model.reactions if rxn.id.endswith('_out')])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fix up those reactions and metabolites that for some reason have a \\_None\\_ suffix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EX_fe2_p_None_:  --> fe2_p_None_\n",
      "EX_ttdca_c_None_: ttdca_c_None_ --> \n",
      "EX_ptdca_c_None_: ptdca_c_None_ --> \n",
      "EX_hpdca_c_None_: hpdca_c_None_ --> \n",
      "EX_ttdcea_c_None_: ttdcea_c_None_ --> \n",
      "EX_cpoa2h_c_None_: cpoa2h_c_None_ --> \n",
      "ttdca_c_None_\n",
      "ptdca_c_None_\n",
      "ttdcea_c_None_\n",
      "hpdca_c_None_\n",
      "cpoa2h_c_None_\n",
      "pc_MC_c_None_\n",
      "pe_MC_c_None_\n",
      "pg_MC_c_None_\n",
      "clpn_MC_c_None_\n",
      "peptido_MC_c_None_\n",
      "colipa_e_None_\n",
      "fe2_p_None_\n",
      "tdecoa_c_None_\n",
      "tdcoa_c_None_\n",
      "ppACP_c_None_\n",
      "ptdcalACP_c_None_\n",
      "hpdcalACP_c_None_\n",
      "cpoa2hcoa_c_None_\n",
      "2ombzl_c_None_\n",
      "2ommbl_c_None_\n",
      "2omhmbl_c_None_\n",
      "Ssq23epx_c_None_\n",
      "codhpre6_c_None_\n",
      "thrp_c_None_\n",
      "4hba_c_None_\n",
      "mc_fattyacid_c_None_\n",
      "mc_fattyacidcoa_c_None_\n",
      "1agpgafa_c_None_\n",
      "pa_MC_c_None_\n",
      "cdpdag_MC_c_None_\n",
      "pgp_MC_c_None_\n",
      "ps_MC_c_None_\n",
      "pme_c_None_\n",
      "pdme_c_None_\n",
      "glyc_c_None_\n",
      "udcpp_c_None_\n",
      "adphep_LD_c_None_\n"
     ]
    }
   ],
   "source": [
    "None_reactions = [rxn for rxn in model.reactions if rxn.id.endswith('_None_')]\n",
    "\n",
    "for rxn in None_reactions:\n",
    "    print rxn\n",
    "    rxn.id = rxn.id.replace('_None_','')\n",
    "    \n",
    "None_metabolites = [met for met in model.metabolites if met.id.endswith('_None_')]\n",
    "\n",
    "for met in None_metabolites:\n",
    "    print met\n",
    "    met.id = met.id.replace('_None_','')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "urea_c_None\n",
      "doxopa_c_None\n",
      "allphn_c_None\n",
      "no3_e_None\n"
     ]
    }
   ],
   "source": [
    "None_reactions = [rxn for rxn in model.reactions if rxn.id.endswith('_None')]\n",
    "\n",
    "for rxn in None_reactions:\n",
    "    print rxn\n",
    "    rxn.id = rxn.id.replace('_None','')\n",
    "    \n",
    "None_metabolites = [met for met in model.metabolites if met.id.endswith('_None')]\n",
    "\n",
    "for met in None_metabolites:\n",
    "    print met\n",
    "    met.id = met.id.replace('_None','')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.15549257432771105"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.solve().f"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add transport reactions based on HMMER prediction and parsing against TCDB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import TMHMM as a pandas dataframe, clean up the 'IDs' which currently are FASTA line markers (>lcl|AE017282.2_prot_MCAXXXX_XXXX), but I want to reduce them to just the gene IDs (MCAXXXX)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from pandas import read_csv\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "filepath = '/Users/clie/Desktop/EFPro2/Experimental Data & Standards/Transmembrane Domain Analysis TMHMM/AE017282.2 - TMHMM result short.csv'\n",
    "TMHMM_raw = read_csv(filepath, delimiter='\\t')\n",
    "TMHMM_raw.ID = TMHMM_raw.ID.apply(lambda x: x.split('_')[-2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import PSORT results as a pandas dataframe, clean up the 'IDs' which currently are FASTA line markers, same as above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "filepath = '/Users/clie/Desktop/EFPro2/Experimental Data & Standards/PsortB - Compartment Prediction/PSORTb.csv'\n",
    "PSORT_raw = read_csv(filepath, delimiter=',')\n",
    "PSORT_raw.SeqID = PSORT_raw.SeqID.apply(lambda x: x.split('_')[-3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reshape the TMHMM_raw pandas dataframe (now with cleaned up IDs) so that it no longer contains:\n",
    "\n",
    "- Rows with an ExpAA value lower than 18. Any value above that and it is very likely to be a transmembrane protein (OR have a signal peptide) according to the [TMHMM 1.0 User Guide.](/Users/clie/Desktop/EFPro2/Experimental Data & Standards/Transmembrane Domain Analysis TMHMM/TMHMM1.0-UserGuide.pdf)\n",
    "\n",
    "- Rows with a PredHel value lower than 1. Any value above that and it is very likely to be a transmembrane protein (OR have a signal peptide) according to the [TMHMM 1.0 User Guide.](/Users/clie/Desktop/EFPro2/Experimental Data & Standards/Transmembrane Domain Analysis TMHMM/TMHMM1.0-UserGuide.pdf)\n",
    "\n",
    "This will give us a list of IDs of putative transporter proteins."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "TMHMM_helix_predicted = TMHMM_raw.loc[(TMHMM_raw['ExpAA'] >= 18) & (TMHMM_raw['PredHel'] >= 1)]\n",
    "# TMHMM_ID2Index = TMHMM_raw.set_index('ID')\n",
    "# TMHMM_dictionary = TMHMM_ID2Index.to_dict('index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "putative_transporter_gene_IDs = list(TMHMM_helix_predicted.set_index('ID').index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we construct the BLAST DB from the TCDB FAA file using Ipython Magic/ Shell commands. First we move the current working directory to the corresponding folder and then run BLAST+'s makeblastdb shell command."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/clie/Desktop/EFPro2/Experimental Data & Standards/TCDB Exports\n"
     ]
    }
   ],
   "source": [
    "%cd '/Users/clie/Desktop/EFPro2/Experimental Data & Standards/TCDB Exports/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if os.path.isfile('TCDBSeqDB.pin'):\n",
    "    pass\n",
    "else:\n",
    "    !makeblastdb -in TCDBSeq.faa -dbtype prot -out TCDBSeqDB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now for each putative transporter gene ID from the TMHMM_helix_predicted Dataframe, lets grab the AA sequence, then compile them all into a FAA file a preparation to BLAST it against the TCDB BLAST DB."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "full_central_map = pd.read_csv('/Users/clie/Desktop/EFPro2/Genome/UniProtExport/FullCentralMappingOriginalGenBank.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "ID_2_AASeq = full_central_map.set_index('Unnamed: 0').to_dict('index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "manual_check_bin = []\n",
    "\n",
    "if os.path.isfile('putative_transporters.faa'):\n",
    "    pass\n",
    "else:\n",
    "    !touch putative_transporters.faa\n",
    "    n=60\n",
    "    for gene_id in putative_transporter_gene_IDs:\n",
    "        try: \n",
    "            sequence = ID_2_AASeq[gene_id]['AA Sequence']\n",
    "            seq_snippets = [sequence[i:i+n] for i in range(0,len(sequence),n)]\n",
    "            FASTA_tag = '\\>' + gene_id\n",
    "            !echo $FASTA_tag >> putative_transporters.faa\n",
    "            for snippet in seq_snippets:\n",
    "                !echo $snippet >> putative_transporters.faa\n",
    "        except:\n",
    "            print (\"No sequence found for {0}\".format(gene_id))\n",
    "            manual_check_bin.append(gene_id)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The FASTA file for our query sequences has been created now, so now we BLAST all the query sequences against the TCDBSeqDB, in order to more clearly identify transporters. We will apply strong similarity criteria for the BLAST operation as to avoid getting (m)any false-positives.\n",
    "\n",
    "The criteria will be:\n",
    "- E-value = E^-9\n",
    "- Max target sequences = 1 i.e. I am chosing to ignore multiple matches as I want to recieve unambiguous mappings.\n",
    "- Max Alignments = 1 i.e. I am chosing to ignore multiple matches as I want to recieve unambiguous mappings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "if os.path.isfile('annotated_putative_transporters.csv'):\n",
    "    pass\n",
    "else:\n",
    "    !blastp -db TCDBSeqDB -query putative_transporters.faa -max_hsps 1 -max_target_seqs 1 -evalue 1e-9 -outfmt 10 -out annotated_putative_transporters.csv "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "column_IDs = ['qseqid','sseqid','pident','length','mismatch','gapopen','qstart','qend','sstart','send','evalue','bitscore','TC_Number']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import the BLAST results back in the form of a Pandas Dataframe.\n",
    "\n",
    "After a manual check of the matches with the three lowest bitscores and the three lowest identities I decided not to further restrict this selection. This is, because even though the identity/ overall score was low, the descriptions or e.coli gene names of the qseqid on Uniprot matched those from the TCDB of the sseqid."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "BLAST_results = pd.read_csv('annotated_putative_transporters.csv', names = column_IDs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "BLAST_results.TC_Number = BLAST_results.sseqid.apply(lambda x: x.split('|')[-1])\n",
    "BLAST_results.sseqid = BLAST_results.sseqid.apply(lambda x: x.split('|')[-2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>TC_Number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>MCA0963</td>\n",
       "      <td>B9DIS8</td>\n",
       "      <td>30.114</td>\n",
       "      <td>352</td>\n",
       "      <td>231</td>\n",
       "      <td>9</td>\n",
       "      <td>89</td>\n",
       "      <td>427</td>\n",
       "      <td>14</td>\n",
       "      <td>363</td>\n",
       "      <td>8.570000e-34</td>\n",
       "      <td>136.0</td>\n",
       "      <td>9.B.97.5.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>MCA2175</td>\n",
       "      <td>V4NQP3</td>\n",
       "      <td>42.241</td>\n",
       "      <td>116</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>116</td>\n",
       "      <td>1.750000e-22</td>\n",
       "      <td>84.0</td>\n",
       "      <td>9.B.72.1.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>MCA2301</td>\n",
       "      <td>A5W2Y3</td>\n",
       "      <td>51.149</td>\n",
       "      <td>174</td>\n",
       "      <td>85</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>180</td>\n",
       "      <td>55</td>\n",
       "      <td>228</td>\n",
       "      <td>2.280000e-54</td>\n",
       "      <td>171.0</td>\n",
       "      <td>9.B.69.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>MCA2302</td>\n",
       "      <td>A5W2Y4</td>\n",
       "      <td>50.000</td>\n",
       "      <td>58</td>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>89</td>\n",
       "      <td>5</td>\n",
       "      <td>62</td>\n",
       "      <td>1.490000e-10</td>\n",
       "      <td>50.8</td>\n",
       "      <td>9.B.69.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MCA0019</td>\n",
       "      <td>D5QHE9</td>\n",
       "      <td>36.364</td>\n",
       "      <td>143</td>\n",
       "      <td>85</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>146</td>\n",
       "      <td>42</td>\n",
       "      <td>178</td>\n",
       "      <td>2.480000e-16</td>\n",
       "      <td>71.2</td>\n",
       "      <td>9.B.62.5.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>MCA2966</td>\n",
       "      <td>V5C7X7</td>\n",
       "      <td>64.681</td>\n",
       "      <td>235</td>\n",
       "      <td>80</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>235</td>\n",
       "      <td>1</td>\n",
       "      <td>232</td>\n",
       "      <td>4.960000e-101</td>\n",
       "      <td>292.0</td>\n",
       "      <td>9.B.43.2.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>MCA1650</td>\n",
       "      <td>D4ZZR3</td>\n",
       "      <td>36.538</td>\n",
       "      <td>208</td>\n",
       "      <td>116</td>\n",
       "      <td>5</td>\n",
       "      <td>355</td>\n",
       "      <td>556</td>\n",
       "      <td>196</td>\n",
       "      <td>393</td>\n",
       "      <td>1.110000e-26</td>\n",
       "      <td>112.0</td>\n",
       "      <td>9.B.34.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>MCA1237</td>\n",
       "      <td>D4ZZR3</td>\n",
       "      <td>35.882</td>\n",
       "      <td>170</td>\n",
       "      <td>101</td>\n",
       "      <td>3</td>\n",
       "      <td>275</td>\n",
       "      <td>439</td>\n",
       "      <td>225</td>\n",
       "      <td>391</td>\n",
       "      <td>4.580000e-23</td>\n",
       "      <td>100.0</td>\n",
       "      <td>9.B.34.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>MCA1583</td>\n",
       "      <td>D4ZZR3</td>\n",
       "      <td>32.948</td>\n",
       "      <td>173</td>\n",
       "      <td>107</td>\n",
       "      <td>2</td>\n",
       "      <td>304</td>\n",
       "      <td>470</td>\n",
       "      <td>220</td>\n",
       "      <td>389</td>\n",
       "      <td>1.280000e-23</td>\n",
       "      <td>101.0</td>\n",
       "      <td>9.B.34.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>MCA1533</td>\n",
       "      <td>D4ZZR3</td>\n",
       "      <td>34.104</td>\n",
       "      <td>173</td>\n",
       "      <td>107</td>\n",
       "      <td>2</td>\n",
       "      <td>522</td>\n",
       "      <td>689</td>\n",
       "      <td>221</td>\n",
       "      <td>391</td>\n",
       "      <td>2.470000e-20</td>\n",
       "      <td>93.6</td>\n",
       "      <td>9.B.34.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>MCA0456</td>\n",
       "      <td>D4ZZR3</td>\n",
       "      <td>33.929</td>\n",
       "      <td>168</td>\n",
       "      <td>95</td>\n",
       "      <td>3</td>\n",
       "      <td>285</td>\n",
       "      <td>443</td>\n",
       "      <td>228</td>\n",
       "      <td>388</td>\n",
       "      <td>1.060000e-20</td>\n",
       "      <td>92.8</td>\n",
       "      <td>9.B.34.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>MCA0504</td>\n",
       "      <td>D4ZZR3</td>\n",
       "      <td>30.108</td>\n",
       "      <td>372</td>\n",
       "      <td>223</td>\n",
       "      <td>10</td>\n",
       "      <td>35</td>\n",
       "      <td>384</td>\n",
       "      <td>39</td>\n",
       "      <td>395</td>\n",
       "      <td>3.570000e-36</td>\n",
       "      <td>135.0</td>\n",
       "      <td>9.B.34.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>MCA0547</td>\n",
       "      <td>D4ZZR3</td>\n",
       "      <td>35.359</td>\n",
       "      <td>181</td>\n",
       "      <td>104</td>\n",
       "      <td>4</td>\n",
       "      <td>769</td>\n",
       "      <td>944</td>\n",
       "      <td>221</td>\n",
       "      <td>393</td>\n",
       "      <td>3.970000e-22</td>\n",
       "      <td>99.4</td>\n",
       "      <td>9.B.34.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>MCA1673</td>\n",
       "      <td>D4ZZR3</td>\n",
       "      <td>35.673</td>\n",
       "      <td>171</td>\n",
       "      <td>101</td>\n",
       "      <td>4</td>\n",
       "      <td>902</td>\n",
       "      <td>1066</td>\n",
       "      <td>225</td>\n",
       "      <td>392</td>\n",
       "      <td>1.150000e-24</td>\n",
       "      <td>107.0</td>\n",
       "      <td>9.B.34.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>MCA2989</td>\n",
       "      <td>P60782</td>\n",
       "      <td>45.789</td>\n",
       "      <td>190</td>\n",
       "      <td>101</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>196</td>\n",
       "      <td>8</td>\n",
       "      <td>195</td>\n",
       "      <td>3.300000e-50</td>\n",
       "      <td>160.0</td>\n",
       "      <td>9.B.31.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>MCA0636</td>\n",
       "      <td>E1P8P0</td>\n",
       "      <td>43.165</td>\n",
       "      <td>139</td>\n",
       "      <td>70</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>138</td>\n",
       "      <td>1</td>\n",
       "      <td>131</td>\n",
       "      <td>2.210000e-27</td>\n",
       "      <td>98.2</td>\n",
       "      <td>9.B.27.3.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>MCA2742</td>\n",
       "      <td>E4PCL6</td>\n",
       "      <td>36.478</td>\n",
       "      <td>159</td>\n",
       "      <td>101</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>159</td>\n",
       "      <td>1</td>\n",
       "      <td>159</td>\n",
       "      <td>3.820000e-31</td>\n",
       "      <td>111.0</td>\n",
       "      <td>9.B.27.2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>MCA0667</td>\n",
       "      <td>P0ABP6</td>\n",
       "      <td>52.217</td>\n",
       "      <td>203</td>\n",
       "      <td>95</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>208</td>\n",
       "      <td>8</td>\n",
       "      <td>210</td>\n",
       "      <td>2.100000e-70</td>\n",
       "      <td>213.0</td>\n",
       "      <td>9.B.27.2.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MCA0055</td>\n",
       "      <td>P0ABP6</td>\n",
       "      <td>28.049</td>\n",
       "      <td>164</td>\n",
       "      <td>113</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>180</td>\n",
       "      <td>33</td>\n",
       "      <td>195</td>\n",
       "      <td>1.670000e-12</td>\n",
       "      <td>66.2</td>\n",
       "      <td>9.B.27.2.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>MCA0957</td>\n",
       "      <td>P22037</td>\n",
       "      <td>41.038</td>\n",
       "      <td>212</td>\n",
       "      <td>110</td>\n",
       "      <td>5</td>\n",
       "      <td>29</td>\n",
       "      <td>233</td>\n",
       "      <td>24</td>\n",
       "      <td>227</td>\n",
       "      <td>9.150000e-37</td>\n",
       "      <td>128.0</td>\n",
       "      <td>9.B.20.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MCA0147</td>\n",
       "      <td>G0CIY2</td>\n",
       "      <td>43.925</td>\n",
       "      <td>321</td>\n",
       "      <td>163</td>\n",
       "      <td>6</td>\n",
       "      <td>156</td>\n",
       "      <td>464</td>\n",
       "      <td>169</td>\n",
       "      <td>484</td>\n",
       "      <td>6.640000e-82</td>\n",
       "      <td>260.0</td>\n",
       "      <td>9.B.18.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>MCA2490</td>\n",
       "      <td>P08550</td>\n",
       "      <td>41.135</td>\n",
       "      <td>141</td>\n",
       "      <td>81</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>156</td>\n",
       "      <td>3</td>\n",
       "      <td>142</td>\n",
       "      <td>2.420000e-30</td>\n",
       "      <td>107.0</td>\n",
       "      <td>9.B.160.1.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>MCA1725</td>\n",
       "      <td>B2GH49</td>\n",
       "      <td>38.690</td>\n",
       "      <td>168</td>\n",
       "      <td>95</td>\n",
       "      <td>4</td>\n",
       "      <td>92</td>\n",
       "      <td>258</td>\n",
       "      <td>216</td>\n",
       "      <td>376</td>\n",
       "      <td>1.180000e-16</td>\n",
       "      <td>79.7</td>\n",
       "      <td>9.B.160.1.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>MCA1599</td>\n",
       "      <td>B2GH49</td>\n",
       "      <td>32.184</td>\n",
       "      <td>174</td>\n",
       "      <td>102</td>\n",
       "      <td>6</td>\n",
       "      <td>136</td>\n",
       "      <td>303</td>\n",
       "      <td>217</td>\n",
       "      <td>380</td>\n",
       "      <td>1.280000e-13</td>\n",
       "      <td>71.2</td>\n",
       "      <td>9.B.160.1.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MCA0102</td>\n",
       "      <td>P0ABH4</td>\n",
       "      <td>29.375</td>\n",
       "      <td>160</td>\n",
       "      <td>111</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>162</td>\n",
       "      <td>3</td>\n",
       "      <td>161</td>\n",
       "      <td>1.220000e-20</td>\n",
       "      <td>82.0</td>\n",
       "      <td>9.B.157.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>MCA0167</td>\n",
       "      <td>A2BTW4</td>\n",
       "      <td>26.415</td>\n",
       "      <td>212</td>\n",
       "      <td>126</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>192</td>\n",
       "      <td>18</td>\n",
       "      <td>224</td>\n",
       "      <td>2.030000e-20</td>\n",
       "      <td>84.3</td>\n",
       "      <td>9.B.150.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>MCA2208</td>\n",
       "      <td>Q6MLN2</td>\n",
       "      <td>39.594</td>\n",
       "      <td>197</td>\n",
       "      <td>115</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>195</td>\n",
       "      <td>17</td>\n",
       "      <td>211</td>\n",
       "      <td>9.390000e-39</td>\n",
       "      <td>132.0</td>\n",
       "      <td>9.B.149.1.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>MCA1133</td>\n",
       "      <td>P0AC78</td>\n",
       "      <td>32.673</td>\n",
       "      <td>303</td>\n",
       "      <td>191</td>\n",
       "      <td>5</td>\n",
       "      <td>36</td>\n",
       "      <td>331</td>\n",
       "      <td>26</td>\n",
       "      <td>322</td>\n",
       "      <td>1.070000e-49</td>\n",
       "      <td>171.0</td>\n",
       "      <td>9.B.146.1.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>MCA2432</td>\n",
       "      <td>P0A6W3</td>\n",
       "      <td>64.444</td>\n",
       "      <td>360</td>\n",
       "      <td>128</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>360</td>\n",
       "      <td>1</td>\n",
       "      <td>360</td>\n",
       "      <td>2.590000e-161</td>\n",
       "      <td>454.0</td>\n",
       "      <td>9.B.146.1.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>MCA0902</td>\n",
       "      <td>G2HLY8</td>\n",
       "      <td>34.821</td>\n",
       "      <td>224</td>\n",
       "      <td>137</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>233</td>\n",
       "      <td>70</td>\n",
       "      <td>284</td>\n",
       "      <td>3.770000e-36</td>\n",
       "      <td>130.0</td>\n",
       "      <td>9.B.146.1.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>MCA1000</td>\n",
       "      <td>D8KC08</td>\n",
       "      <td>51.271</td>\n",
       "      <td>236</td>\n",
       "      <td>100</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>236</td>\n",
       "      <td>14</td>\n",
       "      <td>239</td>\n",
       "      <td>1.700000e-63</td>\n",
       "      <td>198.0</td>\n",
       "      <td>1.A.79.2.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MCA0005</td>\n",
       "      <td>B1JFU8</td>\n",
       "      <td>43.333</td>\n",
       "      <td>90</td>\n",
       "      <td>47</td>\n",
       "      <td>2</td>\n",
       "      <td>37</td>\n",
       "      <td>122</td>\n",
       "      <td>37</td>\n",
       "      <td>126</td>\n",
       "      <td>7.330000e-13</td>\n",
       "      <td>60.1</td>\n",
       "      <td>1.A.77.3.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>MCA1464</td>\n",
       "      <td>Q5ZUD4</td>\n",
       "      <td>30.894</td>\n",
       "      <td>123</td>\n",
       "      <td>79</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>122</td>\n",
       "      <td>2</td>\n",
       "      <td>124</td>\n",
       "      <td>1.260000e-14</td>\n",
       "      <td>64.3</td>\n",
       "      <td>1.A.63.2.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>MCA1980</td>\n",
       "      <td>Q27433</td>\n",
       "      <td>26.923</td>\n",
       "      <td>156</td>\n",
       "      <td>108</td>\n",
       "      <td>4</td>\n",
       "      <td>18</td>\n",
       "      <td>172</td>\n",
       "      <td>137</td>\n",
       "      <td>287</td>\n",
       "      <td>7.950000e-10</td>\n",
       "      <td>58.2</td>\n",
       "      <td>1.A.6.2.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>MCA3057</td>\n",
       "      <td>B2N0W4</td>\n",
       "      <td>33.219</td>\n",
       "      <td>292</td>\n",
       "      <td>175</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>274</td>\n",
       "      <td>18</td>\n",
       "      <td>307</td>\n",
       "      <td>3.040000e-41</td>\n",
       "      <td>143.0</td>\n",
       "      <td>1.A.46.2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>MCA2692</td>\n",
       "      <td>B2N0W4</td>\n",
       "      <td>33.929</td>\n",
       "      <td>280</td>\n",
       "      <td>148</td>\n",
       "      <td>9</td>\n",
       "      <td>36</td>\n",
       "      <td>307</td>\n",
       "      <td>62</td>\n",
       "      <td>312</td>\n",
       "      <td>1.730000e-38</td>\n",
       "      <td>137.0</td>\n",
       "      <td>1.A.46.2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>MCA1973</td>\n",
       "      <td>F0LJH3</td>\n",
       "      <td>45.283</td>\n",
       "      <td>106</td>\n",
       "      <td>56</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>124</td>\n",
       "      <td>15</td>\n",
       "      <td>118</td>\n",
       "      <td>5.590000e-21</td>\n",
       "      <td>83.6</td>\n",
       "      <td>1.A.43.1.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>MCA0631</td>\n",
       "      <td>Q8F192</td>\n",
       "      <td>32.000</td>\n",
       "      <td>200</td>\n",
       "      <td>124</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>194</td>\n",
       "      <td>65</td>\n",
       "      <td>258</td>\n",
       "      <td>4.270000e-30</td>\n",
       "      <td>110.0</td>\n",
       "      <td>1.A.30.2.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>MCA0437</td>\n",
       "      <td>Q8F192</td>\n",
       "      <td>39.189</td>\n",
       "      <td>74</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>66</td>\n",
       "      <td>139</td>\n",
       "      <td>158</td>\n",
       "      <td>231</td>\n",
       "      <td>3.140000e-10</td>\n",
       "      <td>56.2</td>\n",
       "      <td>1.A.30.2.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>MCA0436</td>\n",
       "      <td>Q8F192</td>\n",
       "      <td>35.606</td>\n",
       "      <td>132</td>\n",
       "      <td>85</td>\n",
       "      <td>0</td>\n",
       "      <td>246</td>\n",
       "      <td>377</td>\n",
       "      <td>125</td>\n",
       "      <td>256</td>\n",
       "      <td>1.460000e-16</td>\n",
       "      <td>78.2</td>\n",
       "      <td>1.A.30.2.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>MCA0670</td>\n",
       "      <td>Q1D0D2</td>\n",
       "      <td>44.628</td>\n",
       "      <td>121</td>\n",
       "      <td>66</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>133</td>\n",
       "      <td>13</td>\n",
       "      <td>132</td>\n",
       "      <td>8.540000e-28</td>\n",
       "      <td>99.4</td>\n",
       "      <td>1.A.30.2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>MCA0669</td>\n",
       "      <td>Q1D0D3</td>\n",
       "      <td>34.500</td>\n",
       "      <td>200</td>\n",
       "      <td>117</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>211</td>\n",
       "      <td>36</td>\n",
       "      <td>230</td>\n",
       "      <td>3.940000e-29</td>\n",
       "      <td>107.0</td>\n",
       "      <td>1.A.30.2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>MCA1227</td>\n",
       "      <td>Q1D0D2</td>\n",
       "      <td>40.625</td>\n",
       "      <td>128</td>\n",
       "      <td>70</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>143</td>\n",
       "      <td>13</td>\n",
       "      <td>134</td>\n",
       "      <td>1.010000e-25</td>\n",
       "      <td>94.4</td>\n",
       "      <td>1.A.30.2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>MCA0443</td>\n",
       "      <td>Q1D0D2</td>\n",
       "      <td>44.628</td>\n",
       "      <td>121</td>\n",
       "      <td>66</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>133</td>\n",
       "      <td>13</td>\n",
       "      <td>132</td>\n",
       "      <td>8.540000e-28</td>\n",
       "      <td>99.4</td>\n",
       "      <td>1.A.30.2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>MCA0442</td>\n",
       "      <td>Q1D0D3</td>\n",
       "      <td>33.500</td>\n",
       "      <td>200</td>\n",
       "      <td>119</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "      <td>211</td>\n",
       "      <td>36</td>\n",
       "      <td>230</td>\n",
       "      <td>1.360000e-30</td>\n",
       "      <td>111.0</td>\n",
       "      <td>1.A.30.2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>MCA0438</td>\n",
       "      <td>Q1D0D2</td>\n",
       "      <td>28.448</td>\n",
       "      <td>116</td>\n",
       "      <td>81</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>128</td>\n",
       "      <td>16</td>\n",
       "      <td>130</td>\n",
       "      <td>2.890000e-11</td>\n",
       "      <td>56.6</td>\n",
       "      <td>1.A.30.2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>MCA2237</td>\n",
       "      <td>Q1D0D2</td>\n",
       "      <td>39.831</td>\n",
       "      <td>118</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>129</td>\n",
       "      <td>15</td>\n",
       "      <td>132</td>\n",
       "      <td>1.310000e-21</td>\n",
       "      <td>83.2</td>\n",
       "      <td>1.A.30.2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>MCA0632</td>\n",
       "      <td>Q1D0D2</td>\n",
       "      <td>31.783</td>\n",
       "      <td>129</td>\n",
       "      <td>84</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>140</td>\n",
       "      <td>15</td>\n",
       "      <td>139</td>\n",
       "      <td>4.710000e-14</td>\n",
       "      <td>63.9</td>\n",
       "      <td>1.A.30.2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>MCA2238</td>\n",
       "      <td>P0ABU9</td>\n",
       "      <td>32.051</td>\n",
       "      <td>156</td>\n",
       "      <td>97</td>\n",
       "      <td>4</td>\n",
       "      <td>445</td>\n",
       "      <td>597</td>\n",
       "      <td>71</td>\n",
       "      <td>220</td>\n",
       "      <td>4.390000e-15</td>\n",
       "      <td>73.9</td>\n",
       "      <td>1.A.30.2.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>MCA1226</td>\n",
       "      <td>P0ABU9</td>\n",
       "      <td>53.604</td>\n",
       "      <td>222</td>\n",
       "      <td>103</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>224</td>\n",
       "      <td>2</td>\n",
       "      <td>223</td>\n",
       "      <td>6.190000e-82</td>\n",
       "      <td>243.0</td>\n",
       "      <td>1.A.30.2.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MCA0108</td>\n",
       "      <td>O34442</td>\n",
       "      <td>27.294</td>\n",
       "      <td>425</td>\n",
       "      <td>305</td>\n",
       "      <td>2</td>\n",
       "      <td>57</td>\n",
       "      <td>479</td>\n",
       "      <td>26</td>\n",
       "      <td>448</td>\n",
       "      <td>4.410000e-56</td>\n",
       "      <td>192.0</td>\n",
       "      <td>1.A.26.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>MCA1632</td>\n",
       "      <td>P77338</td>\n",
       "      <td>28.082</td>\n",
       "      <td>292</td>\n",
       "      <td>201</td>\n",
       "      <td>3</td>\n",
       "      <td>485</td>\n",
       "      <td>769</td>\n",
       "      <td>808</td>\n",
       "      <td>1097</td>\n",
       "      <td>6.530000e-33</td>\n",
       "      <td>135.0</td>\n",
       "      <td>1.A.23.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>MCA2608</td>\n",
       "      <td>P77338</td>\n",
       "      <td>34.375</td>\n",
       "      <td>256</td>\n",
       "      <td>168</td>\n",
       "      <td>0</td>\n",
       "      <td>538</td>\n",
       "      <td>793</td>\n",
       "      <td>840</td>\n",
       "      <td>1095</td>\n",
       "      <td>7.880000e-45</td>\n",
       "      <td>173.0</td>\n",
       "      <td>1.A.23.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>MCA1308</td>\n",
       "      <td>P0AAC6</td>\n",
       "      <td>55.556</td>\n",
       "      <td>216</td>\n",
       "      <td>92</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>225</td>\n",
       "      <td>8</td>\n",
       "      <td>219</td>\n",
       "      <td>1.150000e-68</td>\n",
       "      <td>209.0</td>\n",
       "      <td>1.A.14.2.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>MCA3086</td>\n",
       "      <td>F8CM01</td>\n",
       "      <td>33.838</td>\n",
       "      <td>198</td>\n",
       "      <td>122</td>\n",
       "      <td>4</td>\n",
       "      <td>76</td>\n",
       "      <td>264</td>\n",
       "      <td>106</td>\n",
       "      <td>303</td>\n",
       "      <td>1.700000e-19</td>\n",
       "      <td>89.4</td>\n",
       "      <td>1.A.13.4.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>MCA1581</td>\n",
       "      <td>O28528</td>\n",
       "      <td>36.364</td>\n",
       "      <td>418</td>\n",
       "      <td>230</td>\n",
       "      <td>8</td>\n",
       "      <td>42</td>\n",
       "      <td>428</td>\n",
       "      <td>38</td>\n",
       "      <td>450</td>\n",
       "      <td>3.720000e-63</td>\n",
       "      <td>209.0</td>\n",
       "      <td>1.A.11.2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>MCA0268</td>\n",
       "      <td>O67997</td>\n",
       "      <td>46.366</td>\n",
       "      <td>399</td>\n",
       "      <td>209</td>\n",
       "      <td>3</td>\n",
       "      <td>27</td>\n",
       "      <td>424</td>\n",
       "      <td>39</td>\n",
       "      <td>433</td>\n",
       "      <td>7.510000e-115</td>\n",
       "      <td>342.0</td>\n",
       "      <td>1.A.11.1.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>MCA0490</td>\n",
       "      <td>O67997</td>\n",
       "      <td>50.995</td>\n",
       "      <td>402</td>\n",
       "      <td>188</td>\n",
       "      <td>4</td>\n",
       "      <td>28</td>\n",
       "      <td>426</td>\n",
       "      <td>40</td>\n",
       "      <td>435</td>\n",
       "      <td>2.020000e-124</td>\n",
       "      <td>367.0</td>\n",
       "      <td>1.A.11.1.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>MCA2136</td>\n",
       "      <td>P69681</td>\n",
       "      <td>53.636</td>\n",
       "      <td>440</td>\n",
       "      <td>170</td>\n",
       "      <td>4</td>\n",
       "      <td>40</td>\n",
       "      <td>479</td>\n",
       "      <td>22</td>\n",
       "      <td>427</td>\n",
       "      <td>1.010000e-153</td>\n",
       "      <td>443.0</td>\n",
       "      <td>1.A.11.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>MCA0647</td>\n",
       "      <td>M2TPV1</td>\n",
       "      <td>41.322</td>\n",
       "      <td>121</td>\n",
       "      <td>63</td>\n",
       "      <td>3</td>\n",
       "      <td>130</td>\n",
       "      <td>247</td>\n",
       "      <td>111</td>\n",
       "      <td>226</td>\n",
       "      <td>2.800000e-23</td>\n",
       "      <td>93.2</td>\n",
       "      <td>1.A.1.28.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>338 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      qseqid  sseqid  pident  length  mismatch  gapopen  qstart  qend  sstart  \\\n",
       "96   MCA0963  B9DIS8  30.114     352       231        9      89   427      14   \n",
       "242  MCA2175  V4NQP3  42.241     116        67        0       1   116       1   \n",
       "259  MCA2301  A5W2Y3  51.149     174        85        0       7   180      55   \n",
       "260  MCA2302  A5W2Y4  50.000      58        28        1      33    89       5   \n",
       "4    MCA0019  D5QHE9  36.364     143        85        2       4   146      42   \n",
       "320  MCA2966  V5C7X7  64.681     235        80        2       1   235       1   \n",
       "183  MCA1650  D4ZZR3  36.538     208       116        5     355   556     196   \n",
       "134  MCA1237  D4ZZR3  35.882     170       101        3     275   439     225   \n",
       "179  MCA1583  D4ZZR3  32.948     173       107        2     304   470     220   \n",
       "174  MCA1533  D4ZZR3  34.104     173       107        2     522   689     221   \n",
       "41   MCA0456  D4ZZR3  33.929     168        95        3     285   443     228   \n",
       "44   MCA0504  D4ZZR3  30.108     372       223       10      35   384      39   \n",
       "51   MCA0547  D4ZZR3  35.359     181       104        4     769   944     221   \n",
       "186  MCA1673  D4ZZR3  35.673     171       101        4     902  1066     225   \n",
       "323  MCA2989  P60782  45.789     190       101        2       7   196       8   \n",
       "60   MCA0636  E1P8P0  43.165     139        70        3       1   138       1   \n",
       "301  MCA2742  E4PCL6  36.478     159       101        0       1   159       1   \n",
       "63   MCA0667  P0ABP6  52.217     203        95        1       8   208       8   \n",
       "6    MCA0055  P0ABP6  28.049     164       113        3      21   180      33   \n",
       "95   MCA0957  P22037  41.038     212       110        5      29   233      24   \n",
       "19   MCA0147  G0CIY2  43.925     321       163        6     156   464     169   \n",
       "282  MCA2490  P08550  41.135     141        81        2      17   156       3   \n",
       "196  MCA1725  B2GH49  38.690     168        95        4      92   258     216   \n",
       "181  MCA1599  B2GH49  32.184     174       102        6     136   303     217   \n",
       "12   MCA0102  P0ABH4  29.375     160       111        2       4   162       3   \n",
       "24   MCA0167  A2BTW4  26.415     212       126        6       6   192      18   \n",
       "245  MCA2208  Q6MLN2  39.594     197       115        4       1   195      17   \n",
       "118  MCA1133  P0AC78  32.673     303       191        5      36   331      26   \n",
       "277  MCA2432  P0A6W3  64.444     360       128        0       1   360       1   \n",
       "90   MCA0902  G2HLY8  34.821     224       137        2      10   233      70   \n",
       "..       ...     ...     ...     ...       ...      ...     ...   ...     ...   \n",
       "101  MCA1000  D8KC08  51.271     236       100        6       6   236      14   \n",
       "0    MCA0005  B1JFU8  43.333      90        47        2      37   122      37   \n",
       "164  MCA1464  Q5ZUD4  30.894     123        79        3       6   122       2   \n",
       "225  MCA1980  Q27433  26.923     156       108        4      18   172     137   \n",
       "327  MCA3057  B2N0W4  33.219     292       175        4       1   274      18   \n",
       "295  MCA2692  B2N0W4  33.929     280       148        9      36   307      62   \n",
       "224  MCA1973  F0LJH3  45.283     106        56        1      19   124      15   \n",
       "58   MCA0631  Q8F192  32.000     200       124        3       1   194      65   \n",
       "36   MCA0437  Q8F192  39.189      74        45        0      66   139     158   \n",
       "35   MCA0436  Q8F192  35.606     132        85        0     246   377     125   \n",
       "65   MCA0670  Q1D0D2  44.628     121        66        1      13   133      13   \n",
       "64   MCA0669  Q1D0D3  34.500     200       117        3      21   211      36   \n",
       "133  MCA1227  Q1D0D2  40.625     128        70        2      16   143      13   \n",
       "39   MCA0443  Q1D0D2  44.628     121        66        1      13   133      13   \n",
       "38   MCA0442  Q1D0D3  33.500     200       119        3      21   211      36   \n",
       "37   MCA0438  Q1D0D2  28.448     116        81        2      14   128      16   \n",
       "251  MCA2237  Q1D0D2  39.831     118        71        0      12   129      15   \n",
       "59   MCA0632  Q1D0D2  31.783     129        84        3      12   140      15   \n",
       "252  MCA2238  P0ABU9  32.051     156        97        4     445   597      71   \n",
       "132  MCA1226  P0ABU9  53.604     222       103        0       3   224       2   \n",
       "14   MCA0108  O34442  27.294     425       305        2      57   479      26   \n",
       "182  MCA1632  P77338  28.082     292       201        3     485   769     808   \n",
       "289  MCA2608  P77338  34.375     256       168        0     538   793     840   \n",
       "145  MCA1308  P0AAC6  55.556     216        92        3      10   225       8   \n",
       "333  MCA3086  F8CM01  33.838     198       122        4      76   264     106   \n",
       "178  MCA1581  O28528  36.364     418       230        8      42   428      38   \n",
       "29   MCA0268  O67997  46.366     399       209        3      27   424      39   \n",
       "43   MCA0490  O67997  50.995     402       188        4      28   426      40   \n",
       "236  MCA2136  P69681  53.636     440       170        4      40   479      22   \n",
       "61   MCA0647  M2TPV1  41.322     121        63        3     130   247     111   \n",
       "\n",
       "     send         evalue  bitscore     TC_Number  \n",
       "96    363   8.570000e-34     136.0    9.B.97.5.1  \n",
       "242   116   1.750000e-22      84.0    9.B.72.1.6  \n",
       "259   228   2.280000e-54     171.0    9.B.69.1.1  \n",
       "260    62   1.490000e-10      50.8    9.B.69.1.1  \n",
       "4     178   2.480000e-16      71.2    9.B.62.5.2  \n",
       "320   232  4.960000e-101     292.0    9.B.43.2.1  \n",
       "183   393   1.110000e-26     112.0    9.B.34.1.3  \n",
       "134   391   4.580000e-23     100.0    9.B.34.1.3  \n",
       "179   389   1.280000e-23     101.0    9.B.34.1.3  \n",
       "174   391   2.470000e-20      93.6    9.B.34.1.3  \n",
       "41    388   1.060000e-20      92.8    9.B.34.1.3  \n",
       "44    395   3.570000e-36     135.0    9.B.34.1.3  \n",
       "51    393   3.970000e-22      99.4    9.B.34.1.3  \n",
       "186   392   1.150000e-24     107.0    9.B.34.1.3  \n",
       "323   195   3.300000e-50     160.0    9.B.31.1.1  \n",
       "60    131   2.210000e-27      98.2    9.B.27.3.1  \n",
       "301   159   3.820000e-31     111.0    9.B.27.2.7  \n",
       "63    210   2.100000e-70     213.0    9.B.27.2.3  \n",
       "6     195   1.670000e-12      66.2    9.B.27.2.3  \n",
       "95    227   9.150000e-37     128.0    9.B.20.1.1  \n",
       "19    484   6.640000e-82     260.0    9.B.18.1.1  \n",
       "282   142   2.420000e-30     107.0  9.B.160.1.16  \n",
       "196   376   1.180000e-16      79.7  9.B.160.1.10  \n",
       "181   380   1.280000e-13      71.2  9.B.160.1.10  \n",
       "12    161   1.220000e-20      82.0   9.B.157.1.3  \n",
       "24    224   2.030000e-20      84.3   9.B.150.1.3  \n",
       "245   211   9.390000e-39     132.0  9.B.149.1.11  \n",
       "118   322   1.070000e-49     171.0   9.B.146.1.7  \n",
       "277   360  2.590000e-161     454.0   9.B.146.1.6  \n",
       "90    284   3.770000e-36     130.0   9.B.146.1.2  \n",
       "..    ...            ...       ...           ...  \n",
       "101   239   1.700000e-63     198.0    1.A.79.2.1  \n",
       "0     126   7.330000e-13      60.1    1.A.77.3.9  \n",
       "164   124   1.260000e-14      64.3    1.A.63.2.3  \n",
       "225   287   7.950000e-10      58.2     1.A.6.2.2  \n",
       "327   307   3.040000e-41     143.0    1.A.46.2.6  \n",
       "295   312   1.730000e-38     137.0    1.A.46.2.6  \n",
       "224   118   5.590000e-21      83.6    1.A.43.1.7  \n",
       "58    258   4.270000e-30     110.0    1.A.30.2.9  \n",
       "36    231   3.140000e-10      56.2    1.A.30.2.9  \n",
       "35    256   1.460000e-16      78.2    1.A.30.2.9  \n",
       "65    132   8.540000e-28      99.4    1.A.30.2.7  \n",
       "64    230   3.940000e-29     107.0    1.A.30.2.7  \n",
       "133   134   1.010000e-25      94.4    1.A.30.2.7  \n",
       "39    132   8.540000e-28      99.4    1.A.30.2.7  \n",
       "38    230   1.360000e-30     111.0    1.A.30.2.7  \n",
       "37    130   2.890000e-11      56.6    1.A.30.2.7  \n",
       "251   132   1.310000e-21      83.2    1.A.30.2.7  \n",
       "59    139   4.710000e-14      63.9    1.A.30.2.7  \n",
       "252   220   4.390000e-15      73.9    1.A.30.2.2  \n",
       "132   223   6.190000e-82     243.0    1.A.30.2.2  \n",
       "14    448   4.410000e-56     192.0    1.A.26.1.3  \n",
       "182  1097   6.530000e-33     135.0    1.A.23.1.1  \n",
       "289  1095   7.880000e-45     173.0    1.A.23.1.1  \n",
       "145   219   1.150000e-68     209.0    1.A.14.2.1  \n",
       "333   303   1.700000e-19      89.4    1.A.13.4.1  \n",
       "178   450   3.720000e-63     209.0    1.A.11.2.7  \n",
       "29    433  7.510000e-115     342.0    1.A.11.1.4  \n",
       "43    435  2.020000e-124     367.0    1.A.11.1.4  \n",
       "236   427  1.010000e-153     443.0    1.A.11.1.1  \n",
       "61    226   2.800000e-23      93.2    1.A.1.28.4  \n",
       "\n",
       "[338 rows x 13 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BLAST_results.sort_values('TC_Number', ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have the Gene ID matched with corresponding TC-Numbers, we will group genes with identical functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "BLAST_results_refocused = pd.DataFrame([BLAST_results['qseqid'],BLAST_results['TC_Number']]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "TC_number_genes_grouped = BLAST_results_refocused.groupby('TC_Number')\n",
    "TC_to_gene = TC_number_genes_grouped.aggregate(lambda x:list(x))\n",
    "TC_to_gene_dict = TC_to_gene.to_dict('index')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will try to find the TC-Numbers in the curated TCDB-extract from MERLIN (Provided by Bruno and Óscar)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MERLIN_curated_TCDB_export = read_csv('/Users/clie/Desktop/EFPro2/Experimental Data & Standards/TCDB Exports/ MERLIN_metabolites_transport_database-Curated Export.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "TCDB_Dict = MERLIN_curated_TCDB_export.set_index('TCDB ID').to_dict('index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.A.11.2.7 NH3; methylamine in True --\n",
      "3.A.12.1.2 DNA in False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "3.D.9.1.1 proton out True 1: Oxidized donor || 1:Reduced acceptor\n",
      "2.A.7.3.29 drugs out True --\n",
      "3.D.4.10.2 proton:proton:proton:proton out False 2:nitric oxide; 2:Ferrocytochrome c; 2:proton || 1:nitrous oxide; 2:Ferricytochrome c; 1:water\n",
      "9.B.102.5.2 unknown in True --\n",
      "2.A.9.3.1 protein out False --\n",
      "1.A.77.3.9 Mg2+; Ca2+ in False --\n",
      "1.A.23.1.1 K+ out True --\n",
      "3.A.1.102.1 polysaccharide out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "2.A.3.8.12 L-threonine // L-serine in // out True --\n",
      "3.A.1.107.3 heme out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "2.A.1.42.1 lysophospholipid in True --\n",
      "3.A.3.5.18 copper out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "2.A.4.7.7 proton // Mn2+ in // out False --\n",
      "3.A.3.5.20 gold out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "3.A.1.114.1 glycolipids out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "9.B.102.5.5 unknown in True --\n",
      "1.B.52.2.2 unknown in True --\n",
      "5.A.1.6.4 electron:electron out False --\n",
      "3.D.1.9.2 proton:proton:proton:proton out False 1:NADH; 1:ubiquinones; 1:proton || 1:DPN; 1:ubiquinol\n",
      "3.D.1.9.1 proton:proton:proton:proton out False 1:NADH; 1:ubiquinones; 1:proton || 1:DPN; 1:ubiquinol\n",
      "2.A.6.4.1 proton // protein in // out False --\n",
      "3.A.1.11.1 putrescine; spermidine in False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "2.A.1.6.4 proline; glycine betaine; taurine; ectoine; L-Pipecolate; Proline betaine; N,N-dimethylglycine; carnitine; 1-Carboxymethyl-pyridinium : proton; Na+ in:in True --\n",
      "2.A.3.8.28 Serine // Threonine in // out True --\n",
      "9.B.125.2.1 unknown out True --\n",
      "2.A.69.4.3 auxin in False --\n",
      "2.A.49.5.5 chloride ion:chloride ion // proton in // out True --\n",
      "3.A.1.111.2 subtilin out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "2.A.4.1.3 proton:K+ // cadmium; zinc in // out False --\n",
      "2.A.59.1.5 arsenite out False --\n",
      "3.A.15.2.1 pilin out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "1.A.43.1.7 fluoride out True --\n",
      "3.A.1.105.17 drugs out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "9.B.14.2.4 heme out True --\n",
      "3.A.1.1.4 lactose in False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "3.D.1.1.1 proton:proton:proton:proton out False 1:NADH; 1:ubiquinones; 1:proton || 1:DPN; 1:ubiquinol\n",
      "3.D.6.1.1 proton; Na+ in False 1:NADH; 1:oxidized ferredoxin || 1:DPN; 1:reduced ferredoxin\n",
      "2.A.40.4.2 9H-xanthine:proton in:in False --\n",
      "2.A.21.9.2 proline sensor False --\n",
      "2.A.5.5.2 ferrous ion; cobalt ion; Mn2+; cadmium; zinc in False --\n",
      "2.A.20.1.1 orthophosphate:proton in:in False --\n",
      "3.A.3.2.21 Ca2+ // proton in // out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "3.A.5.1.1 protein out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "3.A.1.4.5 urea; thiourea; hydroxyurea in False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "3.A.1.103.1 polysaccharide out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "9.B.69.1.1 cobalt ion in True --\n",
      "2.A.3.3.22 arginine; lysine; histidine; ornithine:proton in:in False --\n",
      "2.A.1.2.60 proton // drugs in // out True --\n",
      "5.A.1.1.1 electron:electron out False --\n",
      "2.A.106.1.4 Ca2+ // proton in // out True --\n",
      "2.A.1.8.2 nitric acid in True --\n",
      "2.A.66.4.1 peptidoglycan in True --\n",
      "3.A.1.125.1 lipoproteins out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "2.A.6.7.5 proton // unknown in // out False --\n",
      "3.A.1.139.3 UDP-D-glucose; ferric iron out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "2.A.1.11.3 oxalate // formate in // out True --\n",
      "3.E.2.1.1 proton:proton out False --\n",
      "2.A.53.3.8 sulfate :proton in:in False --\n",
      "3.A.8.1.1 protein out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "9.B.18.1.1 unknown in True --\n",
      "2.A.53.3.4 molybdenum in True --\n",
      "3.D.3.2.1 proton:proton out False 1:ubiquinol; 2:Ferrocytochrome c || 1:ubiquinones; 2:Ferricytochrome c; 2:proton\n",
      "3.D.4.3.2 proton:proton:proton:proton:proton:proton:proton:proton out False 4:Ferrocytochrome c; 1:O2; 4:proton || 4:Ferricytochrome c; 2:water\n",
      "1.A.14.2.1 Ca2+ in True --\n",
      "2.A.7.1.10 cetyltrimethylammonium bromide; cetylpyridinium chloride; tetraphenylphosphonium; benzalkonium chloride; ethidium bromide; sodium dodecyl sulfate out True --\n",
      "3.A.3.7.2 K+ // Na+ in // out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "1.A.11.1.1 NH3 in True --\n",
      "3.A.1.140.1 -- division False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "1.A.11.1.4 NH3 in True --\n",
      "3.A.3.4.1 nickel; Mg2+ in False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "1.B.22.2.1 protein out False --\n",
      "2.A.86.2.1 AI-2 out True --\n",
      "9.B.27.2.3 selenite in True --\n",
      "2.A.64.1.1 protein out False --\n",
      "5.A.3.2.1 proton // electron:electron in // out False 1:formate; 1:ubiquinone; 1:proton || 1:carbon dioxide; 1:ubiquinol\n",
      "9.B.20.1.1 Mg2+ in True --\n",
      "1.A.30.2.7 proton in True --\n",
      "3.A.3.2.4 Ca2+ // proton in // out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "1.A.30.2.2 proton in True --\n",
      "2.A.1.27.1 3-phenylpropionic acid in True --\n",
      "2.A.50.2.1 alanine in False --\n",
      "2.A.6.8.1 proton // unknown in // out False --\n",
      "9.B.150.1.3 unknown out True --\n",
      "3.A.16.1.1 protein out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "3.A.16.1.2 protein out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "2.A.1.25.2 glycopeptide in True --\n",
      "2.A.28.2.2 pyruvate:Na+ in:in False --\n",
      "2.A.102.4.3 toluene-4-sulfonate in True --\n",
      "3.A.11.2.3 DNA in False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "2.A.102.4.5 4-toluene sulfonate in True --\n",
      "2.A.6.3.7 proton // nodulation factor in // out False --\n",
      "2.A.6.3.6 proton // nodulation factor in // out False --\n",
      "3.A.1.1.34 alpha-1,5-arabinooligosaccharides; alpha-1,5-arabinotriose; alpha-1,5-arabinotetraose in False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "2.A.6.3.1 proton // nodulation factor in // out False --\n",
      "1.A.13.4.1 chloride ion out True --\n",
      "2.A.6.1.5 proton // zinc; cadmium; lead in // out False --\n",
      "2.A.6.1.4 proton // copper; silver metal in // out False --\n",
      "2.A.6.1.6 proton // zinc in // out False --\n",
      "2.A.95.1.5 alanine; asparagine; cysteine; glutamine; glycine; isoleucine; leucine; methionine; phenylalanine; proline; tryptophan; valine; histidine; tyrosine; threonine; serine; selenocysteine out True --\n",
      "2.A.6.1.2 proton //cobalt ion; zinc; cadmium in // out False --\n",
      "3.D.4.2.1 proton:proton:proton:proton:proton:proton:proton:proton out False 4:Ferrocytochrome c; 1:O2; 4:proton || 4:Ferricytochrome c; 2:water\n",
      "9.B.14.1.7 heme out True --\n",
      "3.D.3.3.1 proton:proton out False 1:ubiquinol; 2:Ferrocytochrome c || 1:ubiquinones; 2:Ferricytochrome c; 2:proton\n",
      "3.D.1.2.1 proton:proton:proton:proton out False 1:NADH; 1:ubiquinones; 1:proton || 1:DPN; 1:ubiquinol\n",
      "2.A.6.1.8 proton // copper in // out False --\n",
      "9.A.4.1.1 K+ in False --\n",
      "2.A.6.2.33 proton // fluoroquinolone; microlides; chloramphenicol; tetracycline in // out False --\n",
      "1.E.40.3.1 murein hydrolase out False --\n",
      "9.B.1.1.2 EC:3.4.24.84 enzyme True --\n",
      "2.A.6.1.11 proton // ferrous ion; zinc; Mn2+; cobalt ion; cadmium; copper; nickel; lead in // out False --\n",
      "2.A.37.1.3 proton // K+ in // out True --\n",
      "3.A.1.5.34 oligopeptides in False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "9.A.40.2.2 unknown in True --\n",
      "2.A.99.1.1 nickel in True --\n",
      "3.A.1.7.5 orthophosphate in False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "2.A.72.1.1 K+ in True --\n",
      "2.A.86.1.1 AI-2 out True --\n",
      "3.A.3.7.1 K+ // Na+ in // out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "9.B.27.3.1 unknown in True --\n",
      "3.D.2.1.1 proton in True 1:NADP+; 1:NADH || 1:NADPH; 1:DPN\n",
      "9.B.104.1.2 unknown in True --\n",
      "3.B.1.1.5 Na+ in True 1:D-Methylmalonyl-CoA; 1:proton || 1:propanoyl-CoA; 1:carbon dioxide\n",
      "3.A.1.122.15 heavy metal out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "3.A.1.8.2 molybdate; tungstate in False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "3.A.1.106.2 doxorubicin; verapamil; ethidium; tetraphenylphosphonium; vinblastine out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "3.A.1.106.1 lipid A; glycerophospholipids; azidopine; daunomycin; vinblastine; ethidium out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "2.A.38.1.3 proton:K+ in:in True --\n",
      "2.A.6.2.20 proton // Toxoflavine in // out False --\n",
      "2.A.6.2.22 proton // ethidium; ciprofloxacin; norfloxacin; tetracycline; cefotaxime; rifampicin; erythromycin; chloramphenicol; salicylate in // out False --\n",
      "2.A.6.2.27 proton // triclosan in // out False --\n",
      "3.A.3.5.5 copper; silver metal out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "3.A.2.1.1 proton out False 1:ADP; 1:orthophosphate|| 1:ATP; 1:water\n",
      "2.A.1.3.18 proton // phytoalexins; flavonoids; salicylate in // out True --\n",
      "3.A.3.31.1 transition element cation out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "2.A.1.59.2 unknown in True --\n",
      "3.A.1.203.8 long-chain fatty acyl-CoA in False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "3.D.4.8.1 proton:proton:proton:proton:proton:proton:proton:proton out False 4:Ferrocytochrome c; 1:O2; 4:proton || 4:Ferricytochrome c; 2:water\n",
      "2.A.103.1.2 unknown out False --\n",
      "2.A.103.1.1 unknown out False --\n",
      "1.A.6.2.2 Na+ in True --\n",
      "5.A.2.1.1 electron:electron out False --\n",
      "3.A.15.3.1 protein out False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "3.D.5.1.1 Na+ out False 1:NADH; 1:ubiquinone || 1:DPN; 1:ubiquinol\n",
      "2.A.23.1.7 fumarate; D-malate; L-malate; succinate:proton in:in False --\n",
      "2.A.22.5.3 unknown:Na+ in:in True --\n",
      "2.A.37.4.2 proton // K+ in // out True --\n",
      "9.B.105.1.1 lead out True --\n",
      "3.D.1.6.2 proton:proton:proton:proton out False 1:NADH; 1:ubiquinones; 1:proton || 1:DPN; 1:ubiquinol\n",
      "9.B.31.1.1 unknown in True --\n",
      "2.A.17.4.7 dipeptide; tripeptide:proton in:in False --\n",
      "2.A.19.2.1 proton // ca2+ in // out True --\n",
      "3.D.4.6.1 proton:proton:proton:proton:proton:proton:proton:proton out False 4:Ferrocytochrome c; 1:O2; 4:proton || 4:Ferricytochrome c; 2:water\n",
      "3.D.4.6.2 proton:proton:proton:proton:proton:proton:proton:proton out False 4:Ferrocytochrome c; 1:O2; 4:proton || 4:Ferricytochrome c; 2:water\n",
      "3.D.4.7.1 proton:proton:proton:proton:proton:proton:proton:proton out False 4:Ferrocytochrome c; 1:O2; 4:proton || 4:Ferricytochrome c; 2:water\n",
      "2.A.6.2.16 proton // fluoroquinolone; chloramphenicol; biocides; xenobiotics; chloramphenicol in // out False --\n",
      "2.A.6.2.17 proton // fluoroquinolone; tetracycline; macrolides; chloramphenicol; biocides; triclosan; tetracycline; erythromycin in // out False --\n",
      "3.A.1.5.19 oligopeptides in False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "2.A.19.1.1 proton // ca2+; Na+; K+ in // out True --\n",
      "1.B.42.1.2 lipopolysaccharides out True --\n",
      "3.A.1.27.3 phospholipids in False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "3.A.1.27.1 gamma-hexachlorocyclohexane in False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "3.A.1.6.1 sulfate; thiosulfate in False 1:ATP; 1:water || 1:ADP; 1:orthophosphate\n",
      "8.A.3.3.3 polysaccharides out True --\n",
      "3.A.2.1.2 Na+ out False 1:ADP; 1:orthophosphate|| 1:ATP; 1:water\n",
      "2.A.7.1.3 tetraphenylphosphonium; erythromycin A; ethidium bromide; acriflavine; safranin; Pyronine Y out True --\n"
     ]
    }
   ],
   "source": [
    "non_curated_transporters = []\n",
    "\n",
    "for key in TC_to_gene_dict.keys():\n",
    "    if key in TCDB_Dict.keys():\n",
    "        print key, TCDB_Dict[key]['metabolite'],TCDB_Dict[key]['direction'],TCDB_Dict[key]['reversibility'],TCDB_Dict[key]['reacting_metabolites']\n",
    "    else:\n",
    "        non_curated_transporters.append(key) # Check manually at later stage!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here should now follow resolving the TC Numbers to Reaction strings and translating the metabolite names to metabolite IDs, but as it turned out I had no clearance to use the MERLIN-curated TCDB export. Hence this is postponed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Instead, I will write all entries from the Notes field to the Annotation fields to be included in the most recent version of the SMBL export. Newer, non-legacy formats do not support the Notes field.\n",
    "\n",
    "I will try to match the MIRIAM nomenclature."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for met in model.metabolites:\n",
    "    try:\n",
    "        len(met.notes.keys()) > 1\n",
    "        met_notes = met.notes\n",
    "        anno_dict = {}\n",
    "        for key in met_notes.keys():\n",
    "            if key == 'BIGG':\n",
    "                anno_dict['bigg.metabolite'] = met_notes[key][0].lstrip('\\'').rstrip('\\'').split(',')[0]\n",
    "            elif key == 'KEGG':\n",
    "                    value_entries = met_notes[key][0].lstrip('\\'').rstrip('\\'').split(',')\n",
    "                    if len(value_entries) > 1:\n",
    "                        anno_dict['kegg.compound'] = [x.lstrip(' ') for x in value_entries]\n",
    "                    else:\n",
    "                        anno_dict['kegg.compound'] = value_entries[0].lstrip(' ')\n",
    "            elif key == 'SEED':\n",
    "                value_entries = met_notes[key][0].lstrip('\\'').rstrip('\\'').split(',')\n",
    "                if len(value_entries) > 1:\n",
    "                    anno_dict['seed.compound'] = [x for x in value_entries]\n",
    "                else:\n",
    "                    anno_dict['seed.compound'] = value_entries[0]\n",
    "            elif key == 'MXNREF':\n",
    "                value_entries = met_notes[key][0].lstrip('\\'').rstrip('\\'').split(',')\n",
    "                if len(value_entries) > 1:\n",
    "                    anno_dict['metanetx.chemical'] = [key+':'+x.lstrip(' ') for x in value_entries]\n",
    "                else:\n",
    "                    anno_dict['metanetx.chemical'] = value_entries[0]\n",
    "            elif key == 'METACYC':\n",
    "                value_entries = met_notes[key][0].lstrip('\\'').rstrip('\\'').split(',')\n",
    "                if len(value_entries) > 1:\n",
    "                    anno_dict['biocyc'] = ['META:'+x.lstrip(' ') for x in value_entries]\n",
    "                else:\n",
    "                    anno_dict['biocyc'] = 'META:'+ value_entries[0].lstrip(' ')\n",
    "            elif key == 'SMILES':\n",
    "                pass\n",
    "            elif key == 'BRENDA':\n",
    "                pass\n",
    "            elif key == 'INCHI':\n",
    "                value_entries = met_notes[key][0].lstrip('\\'').rstrip('\\'')\n",
    "                anno_dict['inchi'] = met_notes[key][0].lstrip('\\'').rstrip('\\'')\n",
    "            elif key == 'HMDB':\n",
    "                value_entries = met_notes[key][0].lstrip('\\'').rstrip('\\'').split(',')\n",
    "                if len(value_entries) > 1:\n",
    "                    anno_dict[key.lower()] = [x.lstrip(' ') for x in value_entries]\n",
    "                else:\n",
    "                    anno_dict[key.lower()] = value_entries[0].lstrip(' ')\n",
    "            else:\n",
    "                value_entries = met_notes[key][0].lstrip('\\'').rstrip('\\'').split(',')\n",
    "                if len(value_entries) > 1:\n",
    "                    anno_dict[key.lower()] = [key+':'+x.lstrip(' ') for x in value_entries]\n",
    "                else:\n",
    "                    anno_dict[key.lower()] = key+':'+value_entries[0].lstrip(' ')\n",
    "        met.annotation = anno_dict\n",
    "    except:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for rxn in model.reactions:\n",
    "    try:\n",
    "        len(rxn.notes.keys()) > 1\n",
    "        rxn_notes = rxn.notes\n",
    "        anno_dict = {}\n",
    "        for key in rxn_notes.keys():\n",
    "            if key == 'BIGG':\n",
    "                anno_dict['bigg.reaction'] = rxn_notes[key][0].lstrip('\\'').rstrip('\\'').split(',')[0]\n",
    "            elif key == 'KEGG':\n",
    "                    value_entries = rxn_notes[key][0].lstrip('\\'').rstrip('\\'').split(',')\n",
    "                    if len(value_entries) > 1:\n",
    "                        anno_dict['kegg.reaction'] = [x.lstrip(' ') for x in value_entries]\n",
    "                    else:\n",
    "                        anno_dict['kegg.reaction'] = value_entries[0].lstrip(' ')\n",
    "            elif key == 'SEED':\n",
    "                value_entries = rxn_notes[key][0].lstrip('\\'').rstrip('\\'').split(',')\n",
    "                if len(value_entries) > 1:\n",
    "                    anno_dict['seed'] = [x.lstrip(' ') for x in value_entries]\n",
    "                else:\n",
    "                    anno_dict['seed'] = value_entries[0].lstrip(' ')\n",
    "            elif key == 'MXNREF':\n",
    "                value_entries = rxn_notes[key][0].lstrip('\\'').rstrip('\\'').split(',')\n",
    "                if len(value_entries) > 1:\n",
    "                    anno_dict['metanetx.reaction'] = [key+':'+x.lstrip(' ') for x in value_entries]\n",
    "                else:\n",
    "                    anno_dict['metanetx.reaction'] = value_entries[0]\n",
    "            elif key == 'METACYC':\n",
    "                value_entries = rxn_notes[key][0].lstrip('\\'').rstrip('\\'').split(',')\n",
    "                if len(value_entries) > 1:\n",
    "                    anno_dict['biocyc'] = ['META:'+x.lstrip(' ') for x in value_entries]\n",
    "                else:\n",
    "                    anno_dict['biocyc'] = 'META:'+ value_entries[0].lstrip(' ')\n",
    "            elif key == 'GENE ASSOCIATION':\n",
    "                pass\n",
    "            elif key == 'CONFIDENCE SCORE':\n",
    "                pass\n",
    "            elif key == 'INHIBITOR':\n",
    "                pass\n",
    "            elif key == 'LOCALIZATION':\n",
    "                pass\n",
    "            elif key == 'COFACTOR':\n",
    "                pass\n",
    "            elif key == 'SUBSYSTEM':\n",
    "                pass\n",
    "            elif key == 'BRENDA':\n",
    "                pass\n",
    "            elif key == 'EC NUMBER':\n",
    "                value_entries = rxn_notes[key][0].lstrip('\\'').rstrip('\\'').split(';')\n",
    "                if len(value_entries) > 1:\n",
    "                    anno_dict['brenda'] = [x.lstrip(' ') for x in value_entries]\n",
    "                else:\n",
    "                    anno_dict['brenda'] = value_entries[0].lstrip(' ')\n",
    "            elif key == 'REACTOME':\n",
    "                value_entries = rxn_notes[key][0].lstrip('\\'').rstrip('\\'').split(',')\n",
    "                if len(value_entries) > 1:\n",
    "                    anno_dict['reactome'] = [x.lstrip(' ') for x in value_entries]\n",
    "                else:\n",
    "                    anno_dict['reactome'] = value_entries[0].lstrip(' ')\n",
    "            elif key == 'RHEA':\n",
    "                value_entries = rxn_notes[key][0].lstrip('\\'').rstrip('\\'').split(',')\n",
    "                if len(value_entries) > 1:\n",
    "                    anno_dict['rhea'] = [x.lstrip(' ') for x in value_entries]\n",
    "                else:\n",
    "                    anno_dict['rhea'] = value_entries[0].lstrip(' ')\n",
    "            else:\n",
    "                value_entries = rxn_notes[key][0].lstrip('\\'').rstrip('\\'').split(',')\n",
    "                if len(value_entries) > 1:\n",
    "                    anno_dict[key.lower()] = [key+':'+x.lstrip(' ') for x in value_entries]\n",
    "                else:\n",
    "                    anno_dict[key.lower()] = key+':'+value_entries[0].lstrip(' ')\n",
    "        rxn.annotation = anno_dict\n",
    "    except:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for rxn in model.reactions:\n",
    "    if type(rxn.notes) == dict:\n",
    "        for key in rxn.notes.keys():\n",
    "            if key in ['CONFIDENCE SCORE', 'LOCALIZATION', 'SUBSYSTEM','INHIBITOR','COFACTOR']:\n",
    "                value_entries = rxn.notes[key][0].strip('\\'').split(',')\n",
    "                if len(value_entries) > 1:\n",
    "                    rxn.notes[key] = [x.lstrip(' ') for x in value_entries]\n",
    "                else:\n",
    "                    rxn.notes[key] = value_entries[0]\n",
    "            else:\n",
    "                rxn.notes.pop(key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for met in model.metabolites:\n",
    "    try:\n",
    "        for key in met.notes.keys():\n",
    "            if key not in ['SMILES']:\n",
    "                met.notes.pop(key)\n",
    "    except:\n",
    "        print '{1} has no notes section'.format(met)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on errors I recieved with the SMBL Model Validator, here are some last fixes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Metabolites must be exported with having a compartment at met.compartment!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mets_with_missing_compartments = [met for met in model.metabolites if met.compartment == '']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for met in mets_with_missing_compartments:\n",
    "    met.compartment = met.id.split('_')[-1]\n",
    "    #print met.compartment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All metabolites must have either no formula or a chemically viable one. 'nan' or 'None' or symbols like '(', ')' and '.' are not accepted with fbc:chemicalFormula"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mets_with_faulty_formula = [met for met in model.metabolites if met.formula == 'nan']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for met in mets_with_faulty_formula:\n",
    "    met.formula = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mets_with_faulty_formula = [met for met in model.metabolites if ('.' or '(' or ')') in met.formula]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for met in mets_with_faulty_formula:\n",
    "    met.formula = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EX_co2_c 9.80917347274\n",
      "EX_h2o_c 29.839484494\n",
      "EX_pi_c -0.0742559453479\n",
      "EX_ACP_c 2.23345647532e-17\n",
      "EX_hco3_c 2.30021674414\n",
      "EX_o2_c -27.6123082969\n",
      "EX_trdox_c 4.62223186653e-33\n",
      "EX_so4_c -0.0273340396411\n",
      "EX_h2o2_c 0.00108844802029\n",
      "EX_co_c 4.66477722983e-06\n",
      "EX_2dhp_c 0.045398233971\n",
      "EX_dhpt_c 8.67361737989e-19\n",
      "EX_5prdmbz_c 0.000155492574328\n",
      "EX_no3_e -1.3686798476\n",
      "EX_ch4_p 18.405\n",
      "EX_fe2_p 0.000951614554886\n",
      "BIOMASS_REACTION 0.155492574328\n"
     ]
    }
   ],
   "source": [
    "# verify that it still work as it did before:\n",
    "\n",
    "solution = show_uptake_excretion(model,model.reactions.get_by_id('BIOMASS_REACTION'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Export in the model in the most recent SBML version: Level 3, Version 1 and with fbc_package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "target_filename_json = relative_directory + '/Reconstructions/MethylococcusModel9.json'\n",
    "target_filename_xml = relative_directory + '/Reconstructions/MethylococcusModel9.xml'\n",
    "cobra.io.write_sbml_model(model, target_filename_xml, use_fbc_package=True)\n",
    "cobra.io.save_json_model(model, target_filename_json,pretty = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
